--- a/leaf-server/build.gradle.kts
+++ b/leaf-server/build.gradle.kts
@@ -12,7 +_,7 @@
 val leafMavenPublicUrl = "https://maven.nostal.ink/repository/maven-snapshots/" // Leaf - project setup - Add publish repo
 
 dependencies {
-    mache("io.papermc:mache:1.21.4+build.8")
+    mache("io.papermc:mache:1.21.4+build.7")
     paperclip("cn.dreeam:quantumleaper:1.0.0-SNAPSHOT") // Leaf - project setup - Use own paperclip fork
     testRuntimeOnly("org.junit.platform:junit-platform-launcher")
 }
@@ -54,7 +_,36 @@
         }
     }
 
-    activeFork = leaf
+    // Cartridge start - project setup
+    val cartridge = forks.register("cartridge") {
+        forks = leaf
+        upstream.patchRepo("paperServer") {
+            upstreamRepo = leaf.patchedRepo("paperServer")
+            patchesDir = rootDirectory.dir("cartridge-server/paper-patches")
+            outputDir = rootDirectory.dir("paper-server")
+        }
+        upstream.patchRepo("galeServer") {
+            upstreamRepo = leaf.patchedRepo("galeServer")
+            patchesDir = rootDirectory.dir("cartridge-server/gale-patches")
+            outputDir = rootDirectory.dir("gale-server")
+        }
+        upstream.patchDir("leafServer") {
+            upstreamPath = "leaf-server"
+            excludes = setOf(
+                "src/minecraft",
+                "gale-patches",
+                "paper-patches",
+                "minecraft-patches",
+                "build.gradle.kts",
+                "build.gradle.kts.patch"
+            )
+            patchesDir = rootDirectory.dir("cartridge-server/leaf-patches")
+            outputDir = rootDirectory.dir("leaf-server")
+        }
+    }
+
+    activeFork = cartridge
+    // Cartridge end - project setup
     // Leaf end - project setup
     // Gale end - project setup
 
@@ -146,12 +_,16 @@
         resources { srcDir("../paper-server/src/main/resources") }
         java { srcDir("../gale-server/src/main/java") } // Leaf - project setup
         resources { srcDir("../gale-server/src/main/resources") } // Leaf - project setup
+        java { srcDir("../leaf-server/src/main/java") } // Cartridge - project setup
+        resources { srcDir("../leaf-server/src/main/resources") } // Cartridge - project setup
     }
     test {
         java { srcDir("../paper-server/src/test/java") }
         resources { srcDir("../paper-server/src/test/resources") }
         java { srcDir("../gale-server/src/test/java") } // Leaf - project setup
         resources { srcDir("../gale-server/src/test/resources") } // Leaf - project setup
+        java { srcDir("../leaf-server/src/test/java") } // Cartridge - project setup
+        resources { srcDir("../leaf-server/src/test/resources") } // Cartridge - project setup
     }
 }
 
@@ -176,15 +_,16 @@
 }
 
 dependencies {
-    implementation(project(":leaf-api")) // Gale - project setup - Depend on own API // Leaf - project setup
+
+   implementation(project(":cartridge-api")) // Gale - project setup - Depend on own API // Leaf - project setup // Cartridge - project setup
 
     // Leaf start - Libraries
-    implementation("com.github.thatsmusic99:ConfigurationMaster-API:v2.0.0-rc.3") { // Leaf config
+    implementation("com.github.thatsmusic99:ConfigurationMaster-API:v2.0.0-rc.2") { // Leaf config
         exclude(group = "org.yaml", module = "snakeyaml")
     }
-    implementation("com.github.luben:zstd-jni:1.5.7-1") // LinearPaper
+    implementation("com.github.luben:zstd-jni:1.5.6-9") // LinearPaper
     implementation("org.lz4:lz4-java:1.8.0") // LinearPaper
-    implementation("com.github.ben-manes.caffeine:caffeine:3.2.0")
+    implementation("com.github.ben-manes.caffeine:caffeine:3.1.8")
     // Leaf end - Libraries
 
     implementation("ca.spottedleaf:concurrentutil:0.0.3")
@@ -285,14 +_,14 @@
         val gitBranch = git.exec(providers, "rev-parse", "--abbrev-ref", "HEAD").get().trim()
         attributes(
             "Main-Class" to "org.bukkit.craftbukkit.Main",
-            "Implementation-Title" to "Leaf", // Gale - branding changes // Leaf - Rebrand
+            "Implementation-Title" to "Cartridge", // Gale - branding changes // Leaf - Rebrand
             "Implementation-Version" to implementationVersion,
             "Implementation-Vendor" to date,
-            "Specification-Title" to "Leaf", // Gale - branding changes // Leaf - Rebrand
+            "Specification-Title" to "Cartridge", // Gale - branding changes // Leaf - Rebrand // Cartridge - Rebrand
             "Specification-Version" to project.version,
-            "Specification-Vendor" to "Winds Studio", // Gale - branding changes // Leaf - Rebrand
-            "Brand-Id" to "winds-studio:leaf", // Gale - branding changes // Leaf - Rebrand
-            "Brand-Name" to "Leaf", // Gale - branding changes // Leaf - Rebrand
+            "Specification-Vendor" to "MineInAbyss", // Gale - branding changes // Leaf - Rebrand // Cartridge - Rebrand
+            "Brand-Id" to "mineinabyss:cartridge", // Gale - branding changes // Leaf - Rebrand // Cartridge - Rebrand
+            "Brand-Name" to "Cartridge", // Gale - branding changes // Leaf - Rebrand // Cartridge - Rebrand
             "Build-Number" to (build ?: ""),
             "Build-Time" to buildTime.toString(),
             "Git-Branch" to gitBranch,
