From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jason Penilla <11360596+jpenilla@users.noreply.github.com>
Date: Wed, 11 Dec 2024 19:23:25 -0700
Subject: [PATCH] Block Falldamage tweaks


diff --git a/net/minecraft/DefaultUncaughtExceptionHandler.java b/net/minecraft/DefaultUncaughtExceptionHandler.java
index 6341da454460aa6c00075f82163b97286f760eee..a8bc94121a0209faa8dd293ff3c0862dac280cdb 100644
--- a/net/minecraft/DefaultUncaughtExceptionHandler.java
+++ b/net/minecraft/DefaultUncaughtExceptionHandler.java
@@ -12,6 +12,7 @@ public class DefaultUncaughtExceptionHandler implements UncaughtExceptionHandler
 
     @Override
     public void uncaughtException(Thread thread, Throwable exception) {
+        System.out.println("im a feature");
         this.logger.error("Caught previously unhandled exception :", exception);
     }
 }
diff --git a/net/minecraft/data/tags/VanillaBlockTagsProvider.java b/net/minecraft/data/tags/VanillaBlockTagsProvider.java
index 8829f019eb172234e44d2ae862f0dfe7c7909d83..66e667c29bfb42755d1a9a0ac6debf06041d352c 100644
--- a/net/minecraft/data/tags/VanillaBlockTagsProvider.java
+++ b/net/minecraft/data/tags/VanillaBlockTagsProvider.java
@@ -775,7 +775,7 @@ public class VanillaBlockTagsProvider extends IntrinsicHolderTagsProvider<Block>
                 Blocks.CAVE_VINES,
                 Blocks.CAVE_VINES_PLANT
             );
-        this.tag(BlockTags.FALL_DAMAGE_RESETTING).addTag(BlockTags.CLIMBABLE).add(Blocks.SWEET_BERRY_BUSH, Blocks.COBWEB);
+        this.tag(BlockTags.FALL_DAMAGE_RESETTING);
         this.tag(BlockTags.PIGLIN_REPELLENTS)
             .add(Blocks.SOUL_FIRE)
             .add(Blocks.SOUL_TORCH)
diff --git a/net/minecraft/network/protocol/common/ClientboundUpdateTagsPacket.java b/net/minecraft/network/protocol/common/ClientboundUpdateTagsPacket.java
index 3f22a4f612f0a9d582ad2da3f2be794a4913d08f..d9e2d7d93f2f3c222a298756bc38afa426cbfd58 100644
--- a/net/minecraft/network/protocol/common/ClientboundUpdateTagsPacket.java
+++ b/net/minecraft/network/protocol/common/ClientboundUpdateTagsPacket.java
@@ -1,13 +1,27 @@
 package net.minecraft.network.protocol.common;
 
+import java.util.ArrayList;
+import java.util.List;
 import java.util.Map;
+import java.util.Optional;
+
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+import it.unimi.dsi.fastutil.ints.IntList;
 import net.minecraft.core.Registry;
+import net.minecraft.core.registries.BuiltInRegistries;
+import net.minecraft.core.registries.Registries;
+import net.minecraft.nbt.Tag;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.codec.StreamCodec;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.network.protocol.PacketType;
 import net.minecraft.resources.ResourceKey;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.tags.BlockTags;
 import net.minecraft.tags.TagNetworkSerialization;
+import net.minecraft.world.level.block.Block;
+import org.jetbrains.annotations.Nullable;
 
 public class ClientboundUpdateTagsPacket implements Packet<ClientCommonPacketListener> {
     public static final StreamCodec<FriendlyByteBuf, ClientboundUpdateTagsPacket> STREAM_CODEC = Packet.codec(
@@ -37,6 +51,26 @@ public class ClientboundUpdateTagsPacket implements Packet<ClientCommonPacketLis
         handler.handleUpdateTags(this);
     }
 
+    // Cartridge start
+    @Override
+    public void onPacketDispatch(@Nullable ServerPlayer player) {
+        // if null aka not play-phase return
+        if (player == null) {
+            return;
+        }
+        Map<ResourceLocation, IntList> regTags = Optional.ofNullable(tags.get(BuiltInRegistries.BLOCK.key())).map(n -> n.tags).orElse(null);
+        // If the packet does not contain the block-registry, return as tag was not modified
+        if (regTags == null) {
+            return;
+        }
+        IntList climbableIds = regTags.get(BlockTags.CLIMBABLE.location());
+        // If climbable is not in the packet, return
+        if (climbableIds == null) return;
+        // Set the client-side climbable state based on packet
+        player.hasClientSideClimbable = !climbableIds.isEmpty();
+    }
+    // Cartridge end
+
     public Map<ResourceKey<? extends Registry<?>>, TagNetworkSerialization.NetworkPayload> getTags() {
         return this.tags;
     }
diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index e6414643d4c44bcf22d17bf185cc288d2a4f8399..c1198b35097119ad755b45464c9be500e3a24b81 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -429,6 +429,10 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
     // Paper end - rewrite chunk system
     public boolean hasTickedAtLeastOnceInNewWorld = false; // SparklyPaper - parallel world ticking (fixes bug in DreamResourceReset where the inventory is opened AFTER the player has changed worlds, if you click with the quick tp torch in a chest, because the inventory is opened AFTER the player has teleported)
 
+    // Cartridge start
+    public boolean hasClientSideClimbable = true;
+    // Cartridge end
+
     public ServerPlayer(MinecraftServer server, ServerLevel level, GameProfile gameProfile, ClientInformation clientInformation) {
         super(level, level.getSharedSpawnPos(), level.getSharedSpawnAngle(), gameProfile);
         this.textFilter = server.createTextFilterForPlayer(this);
@@ -1255,9 +1259,11 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
                     return false;
                 }
                 // Purpur end - Minecart settings and WASD controls
-                if (getRootVehicle() instanceof net.minecraft.world.entity.vehicle.Boat && !level().purpurConfig.boatsDoFallDamage) {
-                    return false;
-                }
+                // Cartridge start - boats dont do falldamage
+                //if (getRootVehicle() instanceof net.minecraft.world.entity.vehicle.Boat && !level().purpurConfig.boatsDoFallDamage) {
+                //    return false;
+                //}
+                // Cartridge end
             }
             // Purpur end - Add boat fall damage config
             Entity entity = damageSource.getEntity();
@@ -1750,6 +1756,18 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
             this.spawnExtraParticlesOnFall = false;
         }
 
+        BlockState iblockdata = this.level().getBlockState(pos);
+        if (hasLandedInLiquid()) {
+            if (fallDistance < getMaxFallDistance() && level().getBlockStates(this.getBoundingBox().move(0, -2, 0)).allMatch(b -> !b.getFluidState().isEmpty() || b.isSolid())) onGround = true;
+            else if (fallDistance < getMaxFallDistance() || !iblockdata.getFluidState().isEmpty()) onGround = false;
+            if (onGround && (!iblockdata.getFluidState().isEmpty() || !getInBlockState().getFluidState().isEmpty()) && (isEyeInFluid(FluidTags.WATER))) onGround = false;
+            if (!onGround && iblockdata.isSolid() && iblockdata.getFluidState().isEmpty() && !isEyeInFluid(FluidTags.WATER)) onGround = true;
+
+        } else if (fallDistance >= getMaxFallDistance() && !iblockdata.getFluidState().isEmpty()) {
+            if (this.getDeltaMovement().y() > 9.999999747378752E-6D) resetFallDistance();
+            onGround = true;
+        }
+
         super.checkFallDamage(y, onGround, state, pos);
     }
 
@@ -1758,7 +1776,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
         super.onExplosionHit(entity);
         this.currentImpulseImpactPos = this.position();
         this.currentExplosionCause = entity;
-        this.setIgnoreFallDamageFromCurrentImpulse(entity != null && entity.getType() == EntityType.WIND_CHARGE);
+        //this.setIgnoreFallDamageFromCurrentImpulse(entity != null && entity.getType() == EntityType.WIND_CHARGE); // Cartridge
     }
 
     @Override
diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 1a09ea1953169b72d67126283b5f581c2797506a..d5a1139680069d8b7e8529a4c79754db5e0d3463 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -132,6 +132,8 @@ import net.minecraft.resources.ResourceLocation;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.tags.BlockTags;
+import net.minecraft.tags.FluidTags;
 import net.minecraft.util.FutureChain;
 import net.minecraft.util.Mth;
 import net.minecraft.util.SignatureValidator;
@@ -586,9 +588,15 @@ public class ServerGamePacketListenerImpl
                 d4 = d1 - this.vehicleLastGoodY; // Paper - diff on change, used for checking large move vectors above
                 d5 = d2 - this.vehicleLastGoodZ; // Paper - diff on change, used for checking large move vectors above
                 boolean flag1 = rootVehicle.verticalCollisionBelow;
+                // Cartridge start
                 if (rootVehicle instanceof LivingEntity livingEntity && livingEntity.onClimbable()) {
-                    livingEntity.resetFallDistance();
+                    if (livingEntity.getInBlockState().is(BlockTags.CLIMBABLE) && (!(livingEntity instanceof ServerPlayer serverPlayer) || serverPlayer.hasClientSideClimbable)) {
+                        // If on climbable and the last climbable is not empty, they are currently climbing
+                        // If it is not present, they landed on the ladder, and we want to retain their falldistance
+                        if (livingEntity.getLastClimbablePos().isPresent()) livingEntity.resetFallDistance();
+                    }
                 }
+                // Cartridge end
 
                 rootVehicle.move(MoverType.PLAYER, new Vec3(d3, d4, d5));
                 boolean didCollide = toX != rootVehicle.getX() || toY != rootVehicle.getY() || toZ != rootVehicle.getZ(); // Paper - needed here as the difference in Y can be reset - also note: this is only a guess at whether collisions took place, floating point errors can make this true when it shouldn't be...
@@ -693,6 +701,7 @@ public class ServerGamePacketListenerImpl
                 this.vehicleLastGoodX = rootVehicle.getX();
                 this.vehicleLastGoodY = rootVehicle.getY();
                 this.vehicleLastGoodZ = rootVehicle.getZ();
+                if (rootVehicle.fallDistance > this.player.getMaxFallDistance()) this.player.fallDistance = rootVehicle.fallDistance; // Cartridge
             }
         }
     }
@@ -1648,7 +1657,7 @@ public class ServerGamePacketListenerImpl
                                 if (!this.player.isSpectator() && this.player.isAffectedByBlocks()) this.player.recordMovementThroughBlocks(new Vec3(x, y, z), this.player.position()); // Paper - Do not record movement for vehicles/players unaffected by blocks
                                 this.handlePlayerKnownMovement(vec3);
                                 if (flag) {
-                                    this.player.resetFallDistance();
+                                    //this.player.resetFallDistance(); // Cartridge
                                 }
 
                                 if (packet.isOnGround()
@@ -1673,7 +1682,7 @@ public class ServerGamePacketListenerImpl
                                 this.lastGoodZ = this.player.getZ();
                             } else {
                                 this.internalTeleport(x, y, z, f, f1); // CraftBukkit - SPIGOT-1807: Don't call teleport event, when the client thinks the player is falling, because the chunks are not loaded on the client yet.
-                                this.player.doCheckFallDamage(this.player.getX() - x, this.player.getY() - y, this.player.getZ() - z, packet.isOnGround());
+                                this.player.doCheckFallDamage(this.player.getX() - x, this.player.getY() - y, this.player.getZ() - z, packet.isOnGround() || (this.player.isInWater()  && !this.player.isEyeInFluid(FluidTags.WATER))); // Cartridge - Make water not negate falldamage
                             }
                         }
                     }
diff --git a/net/minecraft/tags/TagNetworkSerialization.java b/net/minecraft/tags/TagNetworkSerialization.java
index fcc5320b82b12ef890934a7afe9dc83a1e1ea7b8..404e3c5ebdabe8c83e02befdae71bf0141e676ad 100644
--- a/net/minecraft/tags/TagNetworkSerialization.java
+++ b/net/minecraft/tags/TagNetworkSerialization.java
@@ -45,7 +45,7 @@ public class TagNetworkSerialization {
         return new TagNetworkSerialization.NetworkPayload(map);
     }
 
-    static <T> TagLoader.LoadResult<T> deserializeTagsFromNetwork(Registry<T> registry, TagNetworkSerialization.NetworkPayload payload) {
+    public static <T> TagLoader.LoadResult<T> deserializeTagsFromNetwork(Registry<T> registry, TagNetworkSerialization.NetworkPayload payload) {
         ResourceKey<? extends Registry<T>> resourceKey = registry.key();
         Map<TagKey<T>, List<Holder<T>>> map = new HashMap<>();
         payload.tags.forEach((resourceLocation, list) -> {
@@ -58,7 +58,7 @@ public class TagNetworkSerialization {
 
     public static final class NetworkPayload {
         public static final TagNetworkSerialization.NetworkPayload EMPTY = new TagNetworkSerialization.NetworkPayload(Map.of());
-        final Map<ResourceLocation, IntList> tags;
+        public final Map<ResourceLocation, IntList> tags;
 
         NetworkPayload(Map<ResourceLocation, IntList> tags) {
             this.tags = tags;
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index df23d80d6b18e900414aa02e5c1812f0a10f0fb7..39ce63e23f6f3250437f462a021bb2cb570dffa2 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -83,6 +83,7 @@ import net.minecraft.world.InteractionResult;
 import net.minecraft.world.Nameable;
 import net.minecraft.world.damagesource.DamageSource;
 import net.minecraft.world.damagesource.DamageSources;
+import net.minecraft.world.entity.ai.attributes.Attributes;
 import net.minecraft.world.entity.item.ItemEntity;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.entity.projectile.Projectile;
@@ -922,7 +923,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
 
         if (this.isInLava()) {
             this.lavaHurt();
-            this.fallDistance *= 0.5F;
+            //this.fallDistance *= 0.5F; // Cartridge
             // CraftBukkit start
         } else {
             this.lastLavaContact = null;
@@ -1188,7 +1189,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                             new ClipContext(this.position(), this.position().add(vec3), ClipContext.Block.FALLDAMAGE_RESETTING, ClipContext.Fluid.WATER, this)
                         );
                     if (blockHitResult.getType() != HitResult.Type.MISS) {
-                        this.resetFallDistance();
+                        //this.resetFallDistance(); // Cartridge - Make water not negate falldamage
                     }
                 }
 
@@ -1347,7 +1348,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
     }
 
     private boolean isStateClimbable(BlockState state) {
-        return state.is(BlockTags.CLIMBABLE) || state.is(Blocks.POWDER_SNOW);
+        return (state.is(BlockTags.CLIMBABLE) && (!(this instanceof ServerPlayer serverPlayer) || serverPlayer.hasClientSideClimbable)) || state.is(Blocks.POWDER_SNOW); // Cartridge
     }
 
     private boolean vibrationAndSoundEffectsFromBlock(BlockPos pos, BlockState state, boolean playStepSound, boolean broadcastGameEvent, Vec3 entityPos) {
@@ -1877,7 +1878,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
     }
 
     protected void checkFallDamage(double y, boolean onGround, BlockState state, BlockPos pos) {
-        if (onGround) {
+        if (onGround || (state.is(BlockTags.CLIMBABLE) && this instanceof ServerPlayer serverPlayer && serverPlayer.hasClientSideClimbable)) { // Cartridge
             if (this.fallDistance > 0.0F) {
                 state.getBlock().fallOn(this.level(), state, pos, this, this.fallDistance);
                 this.level()
@@ -1910,7 +1911,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                 }
             }
 
-            return false;
+            return true; // Cartridge
         }
     }
 
@@ -1974,7 +1975,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                 this.doWaterSplashEffect();
             }
 
-            this.resetFallDistance();
+            if (!(this instanceof ServerPlayer)) this.resetFallDistance(); // Cartridge - Make water not negate falldamage
             this.wasTouchingWater = true;
             this.clearFire();
         } else {
@@ -3161,6 +3162,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                     this.stopRiding();
                 }
 
+                this.causeFallDamage(this.fallDistance, 1.0f, this.damageSources().fall()); // Cartridge
                 this.setPose(Pose.STANDING);
                 this.vehicle = vehicle;
                 this.vehicle.addPassenger(this);
@@ -3174,9 +3176,12 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
         }
     }
 
+    // Cartridge start - Prevent entering vehicle when falling
     protected boolean canRide(Entity vehicle) {
-        return !this.isShiftKeyDown() && this.boardingCooldown <= 0;
+        boolean isSafeFallDistance = (!(this instanceof net.minecraft.world.entity.LivingEntity livingEntity) || this.fallDistance < livingEntity.getAttributeValue(Attributes.SAFE_FALL_DISTANCE));
+        return isSafeFallDistance && !this.isShiftKeyDown() && this.boardingCooldown <= 0;
     }
+    // Cartridge end - Prevent entering vehicle when falling
 
     public void ejectPassengers() {
         for (int i = this.passengers.size() - 1; i >= 0; i--) {
@@ -3751,7 +3756,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
     }
 
     public void makeStuckInBlock(BlockState state, Vec3 motionMultiplier) {
-        this.resetFallDistance();
+        //this.resetFallDistance(); // Cartridge
         this.stuckSpeedMultiplier = motionMultiplier;
     }
 
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 3aedae4c11bd33502e3ac2ba57d4f99c4c900a62..9623d5abdffa775ea31da30bb18086c839883de2 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -391,7 +391,7 @@ public abstract class LivingEntity extends Entity implements Attackable {
             this.updateInWaterStateAndDoWaterCurrentPushing();
         }
 
-        if (this.level() instanceof ServerLevel serverLevel && onGround && this.fallDistance > 0.0F) {
+        if (this.level() instanceof ServerLevel serverLevel && (onGround || isInWater()) && this.fallDistance > 0.0F) { // Cartridge - Make water not negate falldamage
             this.onChangedBlock(serverLevel, pos);
             double attributeValue = this.getAttributeValue(Attributes.SAFE_FALL_DISTANCE);
             if (this.fallDistance > attributeValue && !state.isAir()) {
@@ -410,20 +410,28 @@ public abstract class LivingEntity extends Entity implements Attackable {
                 float f = Mth.ceil(this.fallDistance - attributeValue);
                 double min = Math.min((double)(0.2F + f / 15.0F), 2.5);
                 int i = (int)(150.0 * min);
+                // Cartridge start - only show block-particles when on ground
                 // CraftBukkit start - visiblity api
-                if (this instanceof ServerPlayer) {
-                    serverLevel.sendParticlesSource((ServerPlayer) this, new BlockParticleOption(ParticleTypes.BLOCK, state), false, false, x, y1, z, i, 0.0, 0.0, 0.0, 0.15F);
-                } else {
-                    serverLevel.sendParticles(new BlockParticleOption(ParticleTypes.BLOCK, state), x, y1, z, i, 0.0, 0.0, 0.0, 0.15F);
+                if (onGround) {
+                    if (this instanceof ServerPlayer) {
+                        serverLevel.sendParticlesSource((ServerPlayer) this, new BlockParticleOption(ParticleTypes.BLOCK, state), false, false, x, y1, z, i, 0.0, 0.0, 0.0, 0.15F);
+                    } else {
+                        serverLevel.sendParticles(new BlockParticleOption(ParticleTypes.BLOCK, state), x, y1, z, i, 0.0, 0.0, 0.0, 0.15F);
+                    }
                 }
                 // CraftBukkit end
+                // Cartridge end
             }
         }
 
-        super.checkFallDamage(y, onGround, state, pos);
+        // Cartridge start
+        boolean silentOnClimbable = this.getInBlockState().is(BlockTags.CLIMBABLE) && getLastClimbablePos().isPresent();
+        super.checkFallDamage(y, onGround || (silentOnClimbable && getLastClimbablePos().isEmpty()), state, pos);
+
         if (onGround) {
             this.lastClimbablePos = Optional.empty();
         }
+        // Cartridge end
     }
 
     public final boolean canBreatheUnderwater() {
@@ -2111,6 +2119,11 @@ public abstract class LivingEntity extends Entity implements Attackable {
     // CraftBukkit end
 
     public Optional<BlockPos> getLastClimbablePos() {
+        // Cartridge start - correctly set lastClimbablePos
+        if (this instanceof ServerPlayer serverPlayer && !serverPlayer.hasClientSideClimbable) {
+            this.lastClimbablePos = Optional.empty();
+        }
+        // Cartridge end - correctly set lastClimbablePos
         return this.lastClimbablePos;
     }
 
@@ -2145,7 +2158,9 @@ public abstract class LivingEntity extends Entity implements Attackable {
 
     public Boolean onClimbable(boolean loadChunk) {
         // Gale end - don't load chunks to activate climbing entities
-        if (this.isSpectator()) {
+        if (this instanceof ServerPlayer serverPlayer && !serverPlayer.hasClientSideClimbable) { // Cartridge
+            return false;
+        } else if (this.isSpectator()) {
             return false;
         } else {
             BlockPos blockPos = this.blockPosition();
@@ -3587,7 +3602,7 @@ public abstract class LivingEntity extends Entity implements Attackable {
         AABB boundingBox = this.getBoundingBox();
         Vec3 vec3 = new Vec3(this.xxa, this.yya, this.zza);
         if (this.hasEffect(MobEffects.SLOW_FALLING) || this.hasEffect(MobEffects.LEVITATION)) {
-            this.resetFallDistance();
+            //this.resetFallDistance(); // Cartridge
         }
 
         if (this.getControllingPassenger() instanceof Player player && this.isAlive()) {
@@ -3892,7 +3907,7 @@ public abstract class LivingEntity extends Entity implements Attackable {
         super.rideTick();
         this.oRun = this.run;
         this.run = 0.0F;
-        this.resetFallDistance();
+        //this.resetFallDistance(); // Cartridge
     }
 
     @Override
diff --git a/net/minecraft/world/entity/projectile/ThrownEnderpearl.java b/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
index 3712a2143efe82d41bd62dcec826aede03b6448b..f71d1be0a12ef1170f71de4a42c3d5359a4f4b7d 100644
--- a/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
+++ b/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
@@ -125,6 +125,7 @@ public class ThrownEnderpearl extends ThrowableItemProjectile {
 
                 Vec3 vec3 = this.oldPosition();
                 if (owner instanceof ServerPlayer serverPlayer) {
+                    double originalY = serverPlayer.getY();
                     if (serverPlayer.connection.isAcceptingMessages()) {
                         // Leaf start - SparklyPaper - parallel world ticking mod (handling for pearl teleportation cross-dimension)
                         java.util.function.Consumer<ServerPlayer> teleportPlayerCrossDimensionTask = taskServerPlayer -> {
@@ -156,9 +157,10 @@ public class ThrownEnderpearl extends ThrowableItemProjectile {
                         // );
                         // CraftBukkit end - moved up
                         if (serverPlayer1 != null) {
-                            serverPlayer1.resetFallDistance();
-                            serverPlayer1.resetCurrentImpulseContext();
-                            serverPlayer1.hurtServer(taskServerPlayer.serverLevel(), this.damageSources().enderPearl().eventEntityDamager(this), this.level().purpurConfig.enderPearlDamage); // CraftBukkit // Paper - fix DamageSource API // Purpur - Configurable Ender Pearl damage
+                                // Cartridge start - Make Ender Pearls increase fall-damage
+                            //serverPlayer1.resetFallDistance(); // Cartridge
+                                serverPlayer1.fallDistance += Math.max((float) (originalY - this.getY()), 0f);
+                                serverPlayer1.hurtServer(taskServerPlayer.serverLevel(), this.damageSources().enderPearl().eventEntityDamager(this), serverPlayer1.fallDistance); // CraftBukkit // Paper - fix DamageSource API // Purpur - Configurable Ender Pearl damage
                         }
 
                         this.playSound(serverLevel, vec3);
@@ -174,7 +176,7 @@ public class ThrownEnderpearl extends ThrowableItemProjectile {
                         new TeleportTransition(serverLevel, vec3, owner.getDeltaMovement(), owner.getYRot(), owner.getXRot(), TeleportTransition.DO_NOTHING)
                     );
                     if (entity != null) {
-                        entity.resetFallDistance();
+                        //entity.resetFallDistance(); // Cartridge
                     }
 
                     this.playSound(serverLevel, vec3);
diff --git a/net/minecraft/world/entity/vehicle/AbstractBoat.java b/net/minecraft/world/entity/vehicle/AbstractBoat.java
index b1b312e45ed4514eaa6fb3941af64b641220c5bd..3dbae9b8a757cff32d3bf1d193f3d6f17aed2d88 100644
--- a/net/minecraft/world/entity/vehicle/AbstractBoat.java
+++ b/net/minecraft/world/entity/vehicle/AbstractBoat.java
@@ -822,7 +822,7 @@ public abstract class AbstractBoat extends VehicleEntity implements Leashable {
         this.lastYd = this.getDeltaMovement().y;
         if (!this.isPassenger()) {
             if (onGround) {
-                this.resetFallDistance();
+                return; //this.resetFallDistance(); // Cartridge
             } else if (!this.level().getFluidState(this.blockPosition().below()).is(FluidTags.WATER) && y < 0.0) {
                 this.fallDistance -= (float)y;
             }
diff --git a/net/minecraft/world/item/MaceItem.java b/net/minecraft/world/item/MaceItem.java
index d3eb13865ded58147ea652cc98b37923706ab82f..7d171def97bfd1ada20084828a7a6db07abec7e3 100644
--- a/net/minecraft/world/item/MaceItem.java
+++ b/net/minecraft/world/item/MaceItem.java
@@ -72,7 +72,7 @@ public class MaceItem extends Item {
             attacker.setDeltaMovement(attacker.getDeltaMovement().with(Direction.Axis.Y, 0.01F));
             if (attacker instanceof ServerPlayer serverPlayer) {
                 serverPlayer.currentImpulseImpactPos = this.calculateImpactPosition(serverPlayer);
-                serverPlayer.setIgnoreFallDamageFromCurrentImpulse(true);
+                serverPlayer.setIgnoreFallDamageFromCurrentImpulse(false); // Cartridge - prevent mace from resetting fall-distance
                 serverPlayer.connection.send(new ClientboundSetEntityMotionPacket(serverPlayer));
             }
 
@@ -107,7 +107,7 @@ public class MaceItem extends Item {
     public void postHurtEnemy(ItemStack stack, LivingEntity target, LivingEntity attacker) {
         stack.hurtAndBreak(1, attacker, EquipmentSlot.MAINHAND);
         if (canSmashAttack(attacker)) {
-            attacker.resetFallDistance();
+            //attacker.resetFallDistance(); // Cartridge
         }
     }
 
diff --git a/net/minecraft/world/level/block/BedBlock.java b/net/minecraft/world/level/block/BedBlock.java
index 999abf6a27682eed16273c9a73dc9479fc3008c0..28cf30d5535c0c52ceca0aacf62a6a2870792631 100644
--- a/net/minecraft/world/level/block/BedBlock.java
+++ b/net/minecraft/world/level/block/BedBlock.java
@@ -154,7 +154,7 @@ public class BedBlock extends HorizontalDirectionalBlock implements EntityBlock
     // CraftBukkit end
 
     public static boolean canSetSpawn(Level level) {
-        return level.dimensionType().bedWorks();
+        return false; // Cartridge
     }
 
     private boolean kickVillagerOutOfBed(Level level, BlockPos pos) {
diff --git a/net/minecraft/world/level/block/CoralBlock.java b/net/minecraft/world/level/block/CoralBlock.java
index b747af5f3c65f4b79a304b0e903f7b824fb03d8d..b90e5ca511f09522153bda09866e06cff3d6d8fa 100644
--- a/net/minecraft/world/level/block/CoralBlock.java
+++ b/net/minecraft/world/level/block/CoralBlock.java
@@ -34,17 +34,18 @@ public class CoralBlock extends Block {
         return CODEC;
     }
 
-    @Override
-    protected void tick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
-        if (!this.scanForWater(level, pos)) {
-            // CraftBukkit start
-            if (org.bukkit.craftbukkit.event.CraftEventFactory.callBlockFadeEvent(level, pos, this.deadBlock.defaultBlockState()).isCancelled()) {
-                return;
-            }
-            // CraftBukkit end
-            level.setBlock(pos, this.deadBlock.defaultBlockState(), 2);
-        }
-    }
+    // Cartridge - Disable CoralBlocks from fading outside of water
+    //@Override
+    //protected void tick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
+    //    if (!this.scanForWater(level, pos)) {
+    //        // CraftBukkit start
+    //        if (org.bukkit.craftbukkit.event.CraftEventFactory.callBlockFadeEvent(level, pos, this.deadBlock.defaultBlockState()).isCancelled()) {
+    //            return;
+    //        }
+    //        // CraftBukkit end
+    //        level.setBlock(pos, this.deadBlock.defaultBlockState(), 2);
+    //    }
+    //}
 
     @Override
     protected BlockState updateShape(
diff --git a/net/minecraft/world/level/block/CoralFanBlock.java b/net/minecraft/world/level/block/CoralFanBlock.java
index 9530fa0793ed0904b17f8972d09f98a97b2cecb0..a4636d2f1229d0497f3bc5aa65b6c40ff2a8339a 100644
--- a/net/minecraft/world/level/block/CoralFanBlock.java
+++ b/net/minecraft/world/level/block/CoralFanBlock.java
@@ -35,17 +35,18 @@ public class CoralFanBlock extends BaseCoralFanBlock {
         this.tryScheduleDieTick(state, level, level, level.random, pos);
     }
 
-    @Override
-    protected void tick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
-        if (!scanForWater(state, level, pos)) {
-            // CraftBukkit start
-            if (org.bukkit.craftbukkit.event.CraftEventFactory.callBlockFadeEvent(level, pos, this.deadBlock.defaultBlockState().setValue(CoralFanBlock.WATERLOGGED, false)).isCancelled()) {
-                return;
-            }
-            // CraftBukkit end
-            level.setBlock(pos, this.deadBlock.defaultBlockState().setValue(WATERLOGGED, Boolean.valueOf(false)), 2);
-        }
-    }
+    // Cartridge - Disable CoralBlocks from fading outside of water
+    //@Override
+    //protected void tick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
+    //    if (!scanForWater(state, level, pos)) {
+    //        // CraftBukkit start
+    //        if (org.bukkit.craftbukkit.event.CraftEventFactory.callBlockFadeEvent(level, pos, this.deadBlock.defaultBlockState().setValue(CoralFanBlock.WATERLOGGED, false)).isCancelled()) {
+    //            return;
+    //        }
+    //        // CraftBukkit end
+    //        level.setBlock(pos, this.deadBlock.defaultBlockState().setValue(WATERLOGGED, Boolean.valueOf(false)), 2);
+    //    }
+    //}
 
     @Override
     protected BlockState updateShape(
diff --git a/net/minecraft/world/level/block/CoralPlantBlock.java b/net/minecraft/world/level/block/CoralPlantBlock.java
index d1a4dd8ef1ad9d7fb8ae434fc096a9d5b9c5b69d..c4b5ee75b4a332cf4c64e6ab5e7082bb741fa7b4 100644
--- a/net/minecraft/world/level/block/CoralPlantBlock.java
+++ b/net/minecraft/world/level/block/CoralPlantBlock.java
@@ -40,17 +40,18 @@ public class CoralPlantBlock extends BaseCoralPlantTypeBlock {
         this.tryScheduleDieTick(state, level, level, level.random, pos);
     }
 
-    @Override
-    protected void tick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
-        if (!scanForWater(state, level, pos)) {
-            // CraftBukkit start
-            if (org.bukkit.craftbukkit.event.CraftEventFactory.callBlockFadeEvent(level, pos, this.deadBlock.defaultBlockState().setValue(CoralPlantBlock.WATERLOGGED, false)).isCancelled()) {
-                return;
-            }
-            // CraftBukkit end
-            level.setBlock(pos, this.deadBlock.defaultBlockState().setValue(WATERLOGGED, Boolean.valueOf(false)), 2);
-        }
-    }
+    // Cartridge - Disable CoralBlocks from fading outside of water
+    //@Override
+    //protected void tick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
+    //    if (!scanForWater(state, level, pos)) {
+    //        // CraftBukkit start
+    //        if (org.bukkit.craftbukkit.event.CraftEventFactory.callBlockFadeEvent(level, pos, this.deadBlock.defaultBlockState().setValue(CoralPlantBlock.WATERLOGGED, false)).isCancelled()) {
+    //            return;
+    //        }
+    //        // CraftBukkit end
+    //        level.setBlock(pos, this.deadBlock.defaultBlockState().setValue(WATERLOGGED, Boolean.valueOf(false)), 2);
+    //    }
+    //}
 
     @Override
     protected BlockState updateShape(
diff --git a/net/minecraft/world/level/block/CoralWallFanBlock.java b/net/minecraft/world/level/block/CoralWallFanBlock.java
index 5f05dcda8c6eb869d3d6f80f60d4addd5f3bb988..3e36dd81662b5cc063796546b680859b98af48c8 100644
--- a/net/minecraft/world/level/block/CoralWallFanBlock.java
+++ b/net/minecraft/world/level/block/CoralWallFanBlock.java
@@ -35,17 +35,18 @@ public class CoralWallFanBlock extends BaseCoralWallFanBlock {
         this.tryScheduleDieTick(state, level, level, level.random, pos);
     }
 
-    @Override
-    protected void tick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
-        if (!scanForWater(state, level, pos)) {
-            // CraftBukkit start
-            if (org.bukkit.craftbukkit.event.CraftEventFactory.callBlockFadeEvent(level, pos, this.deadBlock.defaultBlockState().setValue(CoralWallFanBlock.WATERLOGGED, false).setValue(CoralWallFanBlock.FACING, state.getValue(CoralWallFanBlock.FACING))).isCancelled()) {
-                return;
-            }
-            // CraftBukkit end
-            level.setBlock(pos, this.deadBlock.defaultBlockState().setValue(WATERLOGGED, Boolean.valueOf(false)).setValue(FACING, state.getValue(FACING)), 2);
-        }
-    }
+    // Cartridge - Disable CoralBlocks from fading outside of water
+    //@Override
+    //protected void tick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
+    //    if (!scanForWater(state, level, pos)) {
+    //        // CraftBukkit start
+    //        if (org.bukkit.craftbukkit.event.CraftEventFactory.callBlockFadeEvent(level, pos, this.deadBlock.defaultBlockState().setValue(CoralWallFanBlock.WATERLOGGED, false).setValue(CoralWallFanBlock.FACING, state.getValue(CoralWallFanBlock.FACING))).isCancelled()) {
+    //            return;
+    //        }
+    //        // CraftBukkit end
+    //        level.setBlock(pos, this.deadBlock.defaultBlockState().setValue(WATERLOGGED, Boolean.valueOf(false)).setValue(FACING, state.getValue(FACING)), 2);
+    //    }
+    //}
 
     @Override
     protected BlockState updateShape(
diff --git a/net/minecraft/world/level/block/FarmBlock.java b/net/minecraft/world/level/block/FarmBlock.java
index 2ff5457300d66378dbbea492deff0136629bc365..7d7258ebba60908daa92643b4b465b8672cc5808 100644
--- a/net/minecraft/world/level/block/FarmBlock.java
+++ b/net/minecraft/world/level/block/FarmBlock.java
@@ -108,57 +108,58 @@ public class FarmBlock extends Block {
         }
     }
 
-    @Override
-    public void fallOn(Level level, BlockState state, BlockPos pos, Entity entity, float fallDistance) {
-        super.fallOn(level, state, pos, entity, fallDistance); // CraftBukkit - moved here as game rules / events shouldn't affect fall damage.
-        if (level instanceof ServerLevel serverLevel
-            && (serverLevel.purpurConfig.farmlandTrampleHeight >= 0D ? fallDistance >= serverLevel.purpurConfig.farmlandTrampleHeight : level.random.nextFloat() < fallDistance - 0.5F) // // Purpur - Configurable farmland trample height
-            && entity instanceof LivingEntity
-            && (entity instanceof Player || serverLevel.purpurConfig.farmlandBypassMobGriefing ^ serverLevel.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING))
-            && entity.getBbWidth() * entity.getBbWidth() * entity.getBbHeight() > 0.512F) {
-                // CraftBukkit start - Interact soil
-                org.bukkit.event.Cancellable cancellable;
-                if (entity instanceof Player) {
-                    cancellable = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerInteractEvent((Player) entity, org.bukkit.event.block.Action.PHYSICAL, pos, null, null, null);
-                } else {
-                    cancellable = new org.bukkit.event.entity.EntityInteractEvent(entity.getBukkitEntity(), level.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ()));
-                    level.getCraftServer().getPluginManager().callEvent((org.bukkit.event.entity.EntityInteractEvent) cancellable);
-                }
-
-                if (cancellable.isCancelled()) {
-                    return;
-                }
-
-                if (level.purpurConfig.farmlandTramplingDisabled) return; // Purpur - Farmland trampling changes
-                if (level.purpurConfig.farmlandTramplingOnlyPlayers && !(entity instanceof Player)) return; // Purpur - Farmland trampling changes
-                
-                // Purpur start - Ability to re-add farmland mechanics from Alpha
-                if (level.purpurConfig.farmlandAlpha) {
-                    Block block = level.getBlockState(pos.below()).getBlock();
-                    if (block instanceof FenceBlock || block instanceof WallBlock) {
-                        return;
-                    }
-                }
-                // Purpur end - Ability to re-add farmland mechanics from Alpha
-
-                // Purpur start - Farmland trampling changes
-                if (level.purpurConfig.farmlandTramplingFeatherFalling) {
-                    java.util.Iterator<net.minecraft.world.item.ItemStack> armor = ((LivingEntity) entity).getArmorSlots().iterator();
-                    if (armor.hasNext() && net.minecraft.world.item.enchantment.EnchantmentHelper.getItemEnchantmentLevel(net.minecraft.world.item.enchantment.Enchantments.FEATHER_FALLING, armor.next()) >= (int) entity.fallDistance) {
-                        return;
-                    }
-                }
-                // Purpur end - Farmland trampling changes
-
-                if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(entity, pos, Blocks.DIRT.defaultBlockState())) {
-                    return;
-                }
-                // CraftBukkit end
-            turnToDirt(entity, state, level, pos);
-        }
-
-        // super.fallOn(level, state, pos, entity, fallDistance); // CraftBukkit - moved up
-    }
+    // Cartridge - Disable jumping on farmland converting to dirt
+    //@Override
+    //public void fallOn(Level level, BlockState state, BlockPos pos, Entity entity, float fallDistance) {
+    //    super.fallOn(level, state, pos, entity, fallDistance); // CraftBukkit - moved here as game rules / events shouldn't affect fall damage.
+    //    if (level instanceof ServerLevel serverLevel
+    //        && (serverLevel.purpurConfig.farmlandTrampleHeight >= 0D ? fallDistance >= serverLevel.purpurConfig.farmlandTrampleHeight : level.random.nextFloat() < fallDistance - 0.5F) // // Purpur - Configurable farmland trample height
+    //        && entity instanceof LivingEntity
+    //        && (entity instanceof Player || serverLevel.purpurConfig.farmlandBypassMobGriefing ^ serverLevel.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING))
+    //        && entity.getBbWidth() * entity.getBbWidth() * entity.getBbHeight() > 0.512F) {
+    //            // CraftBukkit start - Interact soil
+    //            org.bukkit.event.Cancellable cancellable;
+    //            if (entity instanceof Player) {
+    //                cancellable = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerInteractEvent((Player) entity, org.bukkit.event.block.Action.PHYSICAL, pos, null, null, null);
+    //            } else {
+    //                cancellable = new org.bukkit.event.entity.EntityInteractEvent(entity.getBukkitEntity(), level.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ()));
+    //                level.getCraftServer().getPluginManager().callEvent((org.bukkit.event.entity.EntityInteractEvent) cancellable);
+    //            }
+//
+    //            if (cancellable.isCancelled()) {
+    //                return;
+    //            }
+//
+    //            if (level.purpurConfig.farmlandTramplingDisabled) return; // Purpur - Farmland trampling changes
+    //            if (level.purpurConfig.farmlandTramplingOnlyPlayers && !(entity instanceof Player)) return; // Purpur - Farmland trampling changes
+    //
+    //            // Purpur start - Ability to re-add farmland mechanics from Alpha
+    //            if (level.purpurConfig.farmlandAlpha) {
+    //                Block block = level.getBlockState(pos.below()).getBlock();
+    //                if (block instanceof FenceBlock || block instanceof WallBlock) {
+    //                    return;
+    //                }
+    //            }
+    //            // Purpur end - Ability to re-add farmland mechanics from Alpha
+//
+    //            // Purpur start - Farmland trampling changes
+    //            if (level.purpurConfig.farmlandTramplingFeatherFalling) {
+    //                java.util.Iterator<net.minecraft.world.item.ItemStack> armor = ((LivingEntity) entity).getArmorSlots().iterator();
+    //                if (armor.hasNext() && net.minecraft.world.item.enchantment.EnchantmentHelper.getItemEnchantmentLevel(net.minecraft.world.item.enchantment.Enchantments.FEATHER_FALLING, armor.next()) >= (int) entity.fallDistance) {
+    //                    return;
+    //                }
+    //            }
+    //            // Purpur end - Farmland trampling changes
+//
+    //            if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(entity, pos, Blocks.DIRT.defaultBlockState())) {
+    //                return;
+    //            }
+    //            // CraftBukkit end
+    //        turnToDirt(entity, state, level, pos);
+    //    }
+//
+    //    // super.fallOn(level, state, pos, entity, fallDistance); // CraftBukkit - moved up
+    //}
 
     public static void turnToDirt(@Nullable Entity entity, BlockState state, Level level, BlockPos pos) {
         // CraftBukkit start
diff --git a/net/minecraft/world/level/block/HayBlock.java b/net/minecraft/world/level/block/HayBlock.java
index 406e94e4a88921900f59a3a1b65e74a6482c3b8f..30d6c8ea4e48aee65990f62920a7ce693e3a5519 100644
--- a/net/minecraft/world/level/block/HayBlock.java
+++ b/net/minecraft/world/level/block/HayBlock.java
@@ -23,6 +23,6 @@ public class HayBlock extends RotatedPillarBlock {
 
     @Override
     public void fallOn(Level level, BlockState state, BlockPos pos, Entity entity, float fallDistance) {
-        super.fallOn(level, state, pos, entity, fallDistance); // Purpur - Configurable block fall damage modifiers
+        entity.causeFallDamage(fallDistance, 1.0F, level.damageSources().fall()); // Cartridge
     }
 }
diff --git a/net/minecraft/world/level/block/HoneyBlock.java b/net/minecraft/world/level/block/HoneyBlock.java
index bab3ac2c4be08ea7589752b8472c1e13bcaab76a..b22f12495a8dae37449778460a9e251bbf902d50 100644
--- a/net/minecraft/world/level/block/HoneyBlock.java
+++ b/net/minecraft/world/level/block/HoneyBlock.java
@@ -53,7 +53,7 @@ public class HoneyBlock extends HalfTransparentBlock {
             level.broadcastEntityEvent(entity, (byte)54);
         }
 
-        if (entity.causeFallDamage(fallDistance, 0.2F, level.damageSources().fall())) {
+        if (entity.causeFallDamage(fallDistance, 1.0f, level.damageSources().fall())) { // Cartridge
             entity.playSound(this.soundType.getFallSound(), this.soundType.getVolume() * 0.5F, this.soundType.getPitch() * 0.75F);
         }
     }
@@ -108,7 +108,7 @@ public class HoneyBlock extends HalfTransparentBlock {
             entity.setDeltaMovement(new Vec3(deltaMovement.x, getNewDeltaY(-0.05), deltaMovement.z));
         }
 
-        entity.resetFallDistance();
+        //entity.resetFallDistance(); // Cartridge
     }
 
     private void maybeDoSlideEffects(Level level, Entity entity) {
diff --git a/net/minecraft/world/level/block/IceBlock.java b/net/minecraft/world/level/block/IceBlock.java
index a4d735a4365fdaf9e602315aa1176dfd5db77ff5..652d0d0df852d132c9402c773d663bb870354df3 100644
--- a/net/minecraft/world/level/block/IceBlock.java
+++ b/net/minecraft/world/level/block/IceBlock.java
@@ -52,12 +52,13 @@ public class IceBlock extends HalfTransparentBlock {
         }
     }
 
-    @Override
-    protected void randomTick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
-        if (level.getBrightness(LightLayer.BLOCK, pos) > 11 - state.getLightBlock()) {
-            this.melt(state, level, pos);
-        }
-    }
+    // Cartridge - prevent ice from melting
+    //@Override
+    //protected void randomTick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
+    //    if (level.getBrightness(LightLayer.BLOCK, pos) > 11 - state.getLightBlock()) {
+    //        this.melt(state, level, pos);
+    //    }
+    //}
 
     protected void melt(BlockState state, Level level, BlockPos pos) {
         // CraftBukkit start
diff --git a/net/minecraft/world/level/block/NoteBlock.java b/net/minecraft/world/level/block/NoteBlock.java
index 62db23309d722868e406a52b8e67c23273e36c96..62cfb7a681bb8da7a2e0b18a6593968f23a4753a 100644
--- a/net/minecraft/world/level/block/NoteBlock.java
+++ b/net/minecraft/world/level/block/NoteBlock.java
@@ -70,8 +70,7 @@ public class NoteBlock extends Block {
 
     @Override
     public BlockState getStateForPlacement(BlockPlaceContext context) {
-        if (io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableNoteblockUpdates) return this.defaultBlockState(); // Paper - place without considering instrument
-        return this.setInstrument(context.getLevel(), context.getClickedPos(), this.defaultBlockState());
+        return this.defaultBlockState(); // Cartridge - place without considering instrument
     }
 
     @Override
@@ -85,25 +84,12 @@ public class NoteBlock extends Block {
         BlockState neighborState,
         RandomSource random
     ) {
-        if (io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableNoteblockUpdates) return state; // Paper - prevent noteblock instrument from updating
-        boolean flag = direction.getAxis() == Direction.Axis.Y;
-        return flag
-            ? this.setInstrument(level, pos, state)
-            : super.updateShape(state, level, scheduledTickAccess, pos, direction, neighborPos, neighborState, random);
+        return state; // Cartridge
     }
 
     @Override
     protected void neighborChanged(BlockState state, Level level, BlockPos pos, Block neighborBlock, @Nullable Orientation orientation, boolean movedByPiston) {
-        if (io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableNoteblockUpdates) return; // Paper - prevent noteblock powered-state from updating
-        boolean hasNeighborSignal = level.hasNeighborSignal(pos);
-        if (hasNeighborSignal != state.getValue(POWERED)) {
-            if (hasNeighborSignal) {
-                this.playNote(null, state, level, pos);
-                state = level.getBlockState(pos); // CraftBukkit - SPIGOT-5617: update in case changed in event
-            }
-
-            level.setBlock(pos, state.setValue(POWERED, Boolean.valueOf(hasNeighborSignal)), 3);
-        }
+        if (io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableNoteblockUpdates) return; // Cartridge - prevent noteblock powered-state from updating
     }
 
     private void playNote(@Nullable Entity entity, BlockState state, Level level, BlockPos pos) {
diff --git a/net/minecraft/world/level/block/PowderSnowBlock.java b/net/minecraft/world/level/block/PowderSnowBlock.java
index 4f3646961beb877520e257e11224c3045467d351..baa43da6f723fbb00b1922d3777c21e7291bd6cd 100644
--- a/net/minecraft/world/level/block/PowderSnowBlock.java
+++ b/net/minecraft/world/level/block/PowderSnowBlock.java
@@ -97,6 +97,7 @@ public class PowderSnowBlock extends Block implements BucketPickup {
 
     @Override
     public void fallOn(Level level, BlockState state, BlockPos pos, Entity entity, float fallDistance) {
+        entity.causeFallDamage(fallDistance, 1.0F, entity.damageSources().fall()); // Cartridge
         if (!(fallDistance < 4.0) && entity instanceof LivingEntity livingEntity) {
             LivingEntity.Fallsounds fallSounds = livingEntity.getFallSounds();
             SoundEvent soundEvent = fallDistance < 7.0 ? fallSounds.small() : fallSounds.big();
diff --git a/net/minecraft/world/level/block/SlimeBlock.java b/net/minecraft/world/level/block/SlimeBlock.java
index 3cb84ca78d17bbf5af067e06cbaccaf51da336b7..af4338bff3008072d2249ecc734558f5e746c108 100644
--- a/net/minecraft/world/level/block/SlimeBlock.java
+++ b/net/minecraft/world/level/block/SlimeBlock.java
@@ -25,7 +25,7 @@ public class SlimeBlock extends HalfTransparentBlock {
     @Override
     public void fallOn(Level level, BlockState state, BlockPos pos, Entity entity, float fallDistance) {
         if (!entity.isSuppressingBounce()) {
-            entity.causeFallDamage(fallDistance, 0.0F, level.damageSources().fall());
+            entity.causeFallDamage(fallDistance, 1.0F, level.damageSources().fall()); // Cartridge
         }
     }
 
diff --git a/net/minecraft/world/level/block/SweetBerryBushBlock.java b/net/minecraft/world/level/block/SweetBerryBushBlock.java
index 63ec7a5c3b4a0397ce264ca7769ff4adbbab4ef1..aa7800fe05e48eceb4c5f2aad25ded873c405f13 100644
--- a/net/minecraft/world/level/block/SweetBerryBushBlock.java
+++ b/net/minecraft/world/level/block/SweetBerryBushBlock.java
@@ -78,19 +78,21 @@ public class SweetBerryBushBlock extends BushBlock implements BonemealableBlock
     @Override
     protected void entityInside(BlockState state, Level level, BlockPos pos, Entity entity) {
         if (!new io.papermc.paper.event.entity.EntityInsideBlockEvent(entity.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(level, pos)).callEvent()) { return; } // Paper - Add EntityInsideBlockEvent
-        if (entity instanceof LivingEntity && entity.getType() != EntityType.FOX && entity.getType() != EntityType.BEE) {
-            entity.makeStuckInBlock(state, new Vec3(0.8F, 0.75, 0.8F));
-            if (level instanceof ServerLevel serverLevel && state.getValue(AGE) != 0) {
-                Vec3 vec3 = entity.isControlledByClient() ? entity.getKnownMovement() : entity.oldPosition().subtract(entity.position());
-                if (vec3.horizontalDistanceSqr() > 0.0) {
-                    double abs = Math.abs(vec3.x());
-                    double abs1 = Math.abs(vec3.z());
-                    if (abs >= 0.003F || abs1 >= 0.003F) {
-                        entity.hurtServer(serverLevel, level.damageSources().sweetBerryBush().eventBlockDamager(level, pos), 1.0F); // CraftBukkit
-                    }
-                }
-            }
-        }
+        // Cartridge start - Disable BerryBush from damaging entities
+        //if (entity instanceof LivingEntity && entity.getType() != EntityType.FOX && entity.getType() != EntityType.BEE) {
+        //    entity.makeStuckInBlock(state, new Vec3(0.8F, 0.75, 0.8F));
+        //    if (level instanceof ServerLevel serverLevel && state.getValue(AGE) != 0) {
+        //        Vec3 vec3 = entity.isControlledByClient() ? entity.getKnownMovement() : entity.oldPosition().subtract(entity.position());
+        //        if (vec3.horizontalDistanceSqr() > 0.0) {
+        //            double abs = Math.abs(vec3.x());
+        //            double abs1 = Math.abs(vec3.z());
+        //            if (abs >= 0.003F || abs1 >= 0.003F) {
+        //                entity.hurtServer(serverLevel, level.damageSources().sweetBerryBush().directBlock(level, pos), 1.0F); // CraftBukkit
+        //            }
+        //        }
+        //    }
+        //}
+        // Cartridge end - Disable BerryBush from damaging entities
     }
 
     @Override
diff --git a/net/minecraft/world/level/block/TurtleEggBlock.java b/net/minecraft/world/level/block/TurtleEggBlock.java
index a3a093d95306baac22e5cf720f5b14f733b548d4..ca5c275cede3ef1ea4c80ea4e7dc18968e405560 100644
--- a/net/minecraft/world/level/block/TurtleEggBlock.java
+++ b/net/minecraft/world/level/block/TurtleEggBlock.java
@@ -60,14 +60,15 @@ public class TurtleEggBlock extends Block {
         super.stepOn(level, pos, state, entity);
     }
 
-    @Override
-    public void fallOn(Level level, BlockState state, BlockPos pos, Entity entity, float fallDistance) {
-        if (!(entity instanceof Zombie)) {
-            this.destroyEgg(level, state, pos, entity, 3);
-        }
-
-        super.fallOn(level, state, pos, entity, fallDistance);
-    }
+    // Cartridge - Disable eggs from breaking
+    //@Override
+    //public void fallOn(Level level, BlockState state, BlockPos pos, Entity entity, float fallDistance) {
+    //    if (!(entity instanceof Zombie)) {
+    //        this.destroyEgg(level, state, pos, entity, 3);
+    //    }
+//
+    //    super.fallOn(level, state, pos, entity, fallDistance);
+    //}
 
     private void destroyEgg(Level level, BlockState state, BlockPos pos, Entity entity, int chance) {
         if (state.is(Blocks.TURTLE_EGG)
diff --git a/net/minecraft/world/level/material/FlowingFluid.java b/net/minecraft/world/level/material/FlowingFluid.java
index 01ba0756789fcc61905bbaf1be88b526b76f2490..efd9e43991c7dae17fc047eff3ae7f9fbd474869 100644
--- a/net/minecraft/world/level/material/FlowingFluid.java
+++ b/net/minecraft/world/level/material/FlowingFluid.java
@@ -570,39 +570,40 @@ public abstract class FlowingFluid extends Fluid {
         return this.getTickDelay(level);
     }
 
-    @Override
-    public void tick(ServerLevel level, BlockPos pos, BlockState blockState, FluidState fluidState) {
-        if (!fluidState.isSource()) {
-            FluidState newLiquid = this.getNewLiquid(level, pos, level.getBlockState(pos));
-            int spreadDelay = this.getSpreadDelay(level, pos, fluidState, newLiquid);
-            if (newLiquid.isEmpty()) {
-                fluidState = newLiquid;
-                blockState = Blocks.AIR.defaultBlockState();
-                // CraftBukkit start
-                org.bukkit.event.block.FluidLevelChangeEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callFluidLevelChangeEvent(level, pos, blockState);
-                if (event.isCancelled()) {
-                    return;
-                }
-                blockState = ((org.bukkit.craftbukkit.block.data.CraftBlockData) event.getNewData()).getState();
-                // CraftBukkit end
-                level.setBlock(pos, blockState, 3);
-            } else if (!newLiquid.equals(fluidState)) {
-                fluidState = newLiquid;
-                blockState = newLiquid.createLegacyBlock();
-                // CraftBukkit start
-                org.bukkit.event.block.FluidLevelChangeEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callFluidLevelChangeEvent(level, pos, blockState);
-                if (event.isCancelled()) {
-                    return;
-                }
-                blockState = ((org.bukkit.craftbukkit.block.data.CraftBlockData) event.getNewData()).getState();
-                // CraftBukkit end
-                level.setBlock(pos, blockState, 3);
-                level.scheduleTick(pos, newLiquid.getType(), spreadDelay);
-            }
-        }
-
-        this.spread(level, pos, blockState, fluidState);
-    }
+    // Cartridge - Disable liquid flow
+//    @Override
+//    public void tick(ServerLevel level, BlockPos pos, BlockState blockState, FluidState fluidState) {
+//        if (!fluidState.isSource()) {
+//            FluidState newLiquid = this.getNewLiquid(level, pos, level.getBlockState(pos));
+//            int spreadDelay = this.getSpreadDelay(level, pos, fluidState, newLiquid);
+//            if (newLiquid.isEmpty()) {
+//                fluidState = newLiquid;
+//                blockState = Blocks.AIR.defaultBlockState();
+//                // CraftBukkit start
+//                org.bukkit.event.block.FluidLevelChangeEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callFluidLevelChangeEvent(level, pos, blockState);
+//                if (event.isCancelled()) {
+//                    return;
+//                }
+//                blockState = ((org.bukkit.craftbukkit.block.data.CraftBlockData) event.getNewData()).getState();
+//                // CraftBukkit end
+//                level.setBlock(pos, blockState, 3);
+//            } else if (!newLiquid.equals(fluidState)) {
+//                fluidState = newLiquid;
+//                blockState = newLiquid.createLegacyBlock();
+//                // CraftBukkit start
+//                org.bukkit.event.block.FluidLevelChangeEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callFluidLevelChangeEvent(level, pos, blockState);
+//                if (event.isCancelled()) {
+//                    return;
+//                }
+//                blockState = ((org.bukkit.craftbukkit.block.data.CraftBlockData) event.getNewData()).getState();
+//                // CraftBukkit end
+//                level.setBlock(pos, blockState, 3);
+//                level.scheduleTick(pos, newLiquid.getType(), spreadDelay);
+//            }
+//        }
+//
+//        this.spread(level, pos, blockState, fluidState);
+//    }
 
     protected static int getLegacyLevel(FluidState state) {
         return state.isSource() ? 0 : 8 - Math.min(state.getAmount(), 8) + (state.getValue(FALLING) ? 8 : 0);
