From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Boy <sivertpaulsen2@gmail.com>
Date: Thu, 6 Mar 2025 23:33:08 +0100
Subject: [PATCH] Gamerule Tweaks


diff --git a/net/minecraft/world/level/GameRules.java b/net/minecraft/world/level/GameRules.java
index e4ac7ee4e39b609aa70d6d4b766962a4c942911e..a87f305c90530197dcdc279bc9a75d69cbd2b96b 100644
--- a/net/minecraft/world/level/GameRules.java
+++ b/net/minecraft/world/level/GameRules.java
@@ -56,7 +56,7 @@ public class GameRules {
         "mobGriefing", GameRules.Category.MOBS, GameRules.BooleanValue.create(true)
     );
     public static final GameRules.Key<GameRules.BooleanValue> RULE_KEEPINVENTORY = register(
-        "keepInventory", GameRules.Category.PLAYER, GameRules.BooleanValue.create(false)
+        "keepInventory", GameRules.Category.PLAYER, GameRules.BooleanValue.create(true)
     );
     public static final GameRules.Key<GameRules.BooleanValue> RULE_DOMOBSPAWNING = register(
         "doMobSpawning", GameRules.Category.SPAWNING, GameRules.BooleanValue.create(true)
@@ -138,13 +138,13 @@ public class GameRules {
         "commandModificationBlockLimit", GameRules.Category.MISC, GameRules.IntegerValue.create(32768)
     );
     public static final GameRules.Key<GameRules.BooleanValue> RULE_ANNOUNCE_ADVANCEMENTS = register(
-        "announceAdvancements", GameRules.Category.CHAT, GameRules.BooleanValue.create(true)
+        "announceAdvancements", GameRules.Category.CHAT, GameRules.BooleanValue.create(false)
     );
     public static final GameRules.Key<GameRules.BooleanValue> RULE_DISABLE_RAIDS = register(
-        "disableRaids", GameRules.Category.MOBS, GameRules.BooleanValue.create(false)
+        "disableRaids", GameRules.Category.MOBS, GameRules.BooleanValue.create(true)
     );
     public static final GameRules.Key<GameRules.BooleanValue> RULE_DOINSOMNIA = register(
-        "doInsomnia", GameRules.Category.SPAWNING, GameRules.BooleanValue.create(true)
+        "doInsomnia", GameRules.Category.SPAWNING, GameRules.BooleanValue.create(false)
     );
     public static final GameRules.Key<GameRules.BooleanValue> RULE_DO_IMMEDIATE_RESPAWN = register(
         "doImmediateRespawn", GameRules.Category.PLAYER, GameRules.BooleanValue.create(false, (level, value) -> { // Paper - rename param to match changes
@@ -172,13 +172,13 @@ public class GameRules {
         "freezeDamage", GameRules.Category.PLAYER, GameRules.BooleanValue.create(true)
     );
     public static final GameRules.Key<GameRules.BooleanValue> RULE_DO_PATROL_SPAWNING = register(
-        "doPatrolSpawning", GameRules.Category.SPAWNING, GameRules.BooleanValue.create(true)
+        "doPatrolSpawning", GameRules.Category.SPAWNING, GameRules.BooleanValue.create(false)
     );
     public static final GameRules.Key<GameRules.BooleanValue> RULE_DO_TRADER_SPAWNING = register(
-        "doTraderSpawning", GameRules.Category.SPAWNING, GameRules.BooleanValue.create(true)
+        "doTraderSpawning", GameRules.Category.SPAWNING, GameRules.BooleanValue.create(false)
     );
     public static final GameRules.Key<GameRules.BooleanValue> RULE_DO_WARDEN_SPAWNING = register(
-        "doWardenSpawning", GameRules.Category.SPAWNING, GameRules.BooleanValue.create(true)
+        "doWardenSpawning", GameRules.Category.SPAWNING, GameRules.BooleanValue.create(false)
     );
     public static final GameRules.Key<GameRules.BooleanValue> RULE_FORGIVE_DEAD_PLAYERS = register(
         "forgiveDeadPlayers", GameRules.Category.MOBS, GameRules.BooleanValue.create(true)
@@ -190,10 +190,10 @@ public class GameRules {
         "playersSleepingPercentage", GameRules.Category.PLAYER, GameRules.IntegerValue.create(100)
     );
     public static final GameRules.Key<GameRules.BooleanValue> RULE_BLOCK_EXPLOSION_DROP_DECAY = register(
-        "blockExplosionDropDecay", GameRules.Category.DROPS, GameRules.BooleanValue.create(true)
+        "blockExplosionDropDecay", GameRules.Category.DROPS, GameRules.BooleanValue.create(false)
     );
     public static final GameRules.Key<GameRules.BooleanValue> RULE_MOB_EXPLOSION_DROP_DECAY = register(
-        "mobExplosionDropDecay", GameRules.Category.DROPS, GameRules.BooleanValue.create(true)
+        "mobExplosionDropDecay", GameRules.Category.DROPS, GameRules.BooleanValue.create(false)
     );
     public static final GameRules.Key<GameRules.BooleanValue> RULE_TNT_EXPLOSION_DROP_DECAY = register(
         "tntExplosionDropDecay", GameRules.Category.DROPS, GameRules.BooleanValue.create(false)
@@ -211,7 +211,7 @@ public class GameRules {
         "globalSoundEvents", GameRules.Category.MISC, GameRules.BooleanValue.create(true)
     );
     public static final GameRules.Key<GameRules.BooleanValue> RULE_DO_VINES_SPREAD = register(
-        "doVinesSpread", GameRules.Category.UPDATES, GameRules.BooleanValue.create(true)
+        "doVinesSpread", GameRules.Category.UPDATES, GameRules.BooleanValue.create(false)
     );
     public static final GameRules.Key<GameRules.BooleanValue> RULE_ENDER_PEARLS_VANISH_ON_DEATH = register(
         "enderPearlsVanishOnDeath", GameRules.Category.PLAYER, GameRules.BooleanValue.create(true)
