From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Boy <sivertpaulsen2@gmail.com>
Date: Thu, 6 Mar 2025 23:47:00 +0100
Subject: [PATCH] Disable unwanted Allay logic


diff --git a/net/minecraft/world/entity/animal/allay/Allay.java b/net/minecraft/world/entity/animal/allay/Allay.java
index 69d8a10f524dd4db683a00f194e8d1a26c996f43..f598a7eb2ac005da94c71d9664183519cc28f53d 100644
--- a/net/minecraft/world/entity/animal/allay/Allay.java
+++ b/net/minecraft/world/entity/animal/allay/Allay.java
@@ -16,6 +16,7 @@ import net.minecraft.network.syncher.EntityDataAccessor;
 import net.minecraft.network.syncher.EntityDataSerializers;
 import net.minecraft.network.syncher.SynchedEntityData;
 import net.minecraft.server.level.ServerLevel;
+import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.sounds.SoundEvent;
 import net.minecraft.sounds.SoundEvents;
 import net.minecraft.sounds.SoundSource;
@@ -69,7 +70,7 @@ import net.minecraft.world.level.storage.ValueOutput;
 import net.minecraft.world.phys.Vec3;
 import org.jspecify.annotations.Nullable;
 
-public class Allay extends PathfinderMob implements InventoryCarrier, VibrationSystem {
+public class Allay extends PathfinderMob implements InventoryCarrier/*, VibrationSystem*/ {
     private static final Vec3i ITEM_PICKUP_REACH = new Vec3i(1, 1, 1);
     private static final int LIFTING_ITEM_ANIMATION_DURATION = 5;
     private static final float DANCING_LOOP_DURATION = 55.0F;
@@ -88,10 +89,10 @@ public class Allay extends PathfinderMob implements InventoryCarrier, VibrationS
     public static final ImmutableList<Float> THROW_SOUND_PITCHES = ImmutableList.of(
         0.5625F, 0.625F, 0.75F, 0.9375F, 1.0F, 1.0F, 1.125F, 1.25F, 1.5F, 1.875F, 2.0F, 2.25F, 2.5F, 3.0F, 3.75F, 4.0F
     );
-    private final DynamicGameEventListener<VibrationSystem.Listener> dynamicVibrationListener;
-    private VibrationSystem.Data vibrationData;
-    private final VibrationSystem.User vibrationUser;
-    private final DynamicGameEventListener<Allay.JukeboxListener> dynamicJukeboxListener;
+    //private final DynamicGameEventListener<VibrationSystem.Listener> dynamicVibrationListener;
+    //private VibrationSystem.Data vibrationData;
+    //private final VibrationSystem.User vibrationUser;
+    //private final DynamicGameEventListener<Allay.JukeboxListener> dynamicJukeboxListener;
     private final SimpleContainer inventory = new SimpleContainer(1);
     public @Nullable BlockPos jukeboxPos;
     public long duplicationCooldown = 0L;
@@ -105,18 +106,18 @@ public class Allay extends PathfinderMob implements InventoryCarrier, VibrationS
     public Allay(final EntityType<? extends Allay> type, final Level level) {
         super(type, level);
         this.moveControl = new FlyingMoveControl<>(this, 20, true);
-        this.setCanPickUpLoot(this.canPickUpLoot());
-        this.vibrationUser = new Allay.VibrationUser();
-        this.vibrationData = new VibrationSystem.Data();
-        this.dynamicVibrationListener = new DynamicGameEventListener<>(new VibrationSystem.Listener(this));
-        this.dynamicJukeboxListener = new DynamicGameEventListener<>(
-            new Allay.JukeboxListener(this.vibrationUser.getPositionSource(), GameEvent.JUKEBOX_PLAY.value().notificationRadius())
-        );
+        this.setCanPickUpLoot(false); // Cartridge
+//        this.vibrationUser = new Allay.VibrationUser();
+//        this.vibrationData = new VibrationSystem.Data();
+//        this.dynamicVibrationListener = new DynamicGameEventListener<>(new VibrationSystem.Listener(this));
+//        this.dynamicJukeboxListener = new DynamicGameEventListener<>(
+//            new Allay.JukeboxListener(this.vibrationUser.getPositionSource(), GameEvent.JUKEBOX_PLAY.value().notificationRadius())
+//        );
     }
 
     // CraftBukkit start
     public void setCanDuplicate(boolean canDuplicate) {
-        this.entityData.set(Allay.DATA_CAN_DUPLICATE, canDuplicate);
+        this.entityData.set(Allay.DATA_CAN_DUPLICATE, false);
     }
     // CraftBukkit end
 
@@ -151,7 +152,7 @@ public class Allay extends PathfinderMob implements InventoryCarrier, VibrationS
     protected void defineSynchedData(final SynchedEntityData.Builder entityData) {
         super.defineSynchedData(entityData);
         entityData.define(DATA_DANCING, false);
-        entityData.define(DATA_CAN_DUPLICATE, true);
+        entityData.define(DATA_CAN_DUPLICATE, false); // Cartridge
     }
 
     @Override
@@ -221,14 +222,14 @@ public class Allay extends PathfinderMob implements InventoryCarrier, VibrationS
     @Override
     public void aiStep() {
         super.aiStep();
-        if (!this.level().isClientSide() && this.isAlive() && this.tickCount % 10 == 0) {
-            this.heal(1.0F, org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason.REGEN); // CraftBukkit
-        }
+        //if (!this.level().isClientSide() && this.isAlive() && this.tickCount % 10 == 0) {
+        //    this.heal(1.0F, org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason.REGEN); // CraftBukkit
+        //}
 
-        if (this.isDancing() && this.shouldStopDancing() && this.tickCount % 20 == 0) {
-            this.setDancing(false);
-            this.jukeboxPos = null;
-        }
+        //if (this.isDancing() && this.shouldStopDancing() && this.tickCount % 20 == 0) {
+        //    this.setDancing(false);
+        //    this.jukeboxPos = null;
+        //}
 
         this.updateDuplicationCooldown();
     }
@@ -260,7 +261,7 @@ public class Allay extends PathfinderMob implements InventoryCarrier, VibrationS
                 this.spinningAnimationTicks0 = 0.0F;
             }
         } else {
-            VibrationSystem.Ticker.tick(this.level(), this.vibrationData, this.vibrationUser);
+            //VibrationSystem.Ticker.tick(this.level(), this.vibrationData, this.vibrationUser);
             if (this.isPanicking()) {
                 this.setDancing(false);
             }
@@ -269,7 +270,7 @@ public class Allay extends PathfinderMob implements InventoryCarrier, VibrationS
 
     @Override
     public boolean canPickUpLoot() {
-        return !this.isOnPickupCooldown() && this.hasItemInHand();
+        return false;//!this.isOnPickupCooldown() && this.hasItemInHand();
     }
 
     public boolean hasItemInHand() {
@@ -291,53 +292,62 @@ public class Allay extends PathfinderMob implements InventoryCarrier, VibrationS
         ItemStack itemInHand = this.getItemInHand(InteractionHand.MAIN_HAND);
         if (this.isDancing() && interactionItem.is(ItemTags.DUPLICATES_ALLAYS) && this.canDuplicate()) {
             // CraftBukkit start - handle cancel duplication
-            Allay allay = this.duplicateAllay();
-            if (allay == null) {
-                return InteractionResult.SUCCESS;
-            }
+            //Allay allay = this.duplicateAllay();
+            //if (allay == null) {
+            //    return InteractionResult.SUCCESS;
+            //}
             // CraftBukkit end
-            this.level().broadcastEntityEvent(this, EntityEvent.IN_LOVE_HEARTS);
-            this.level().playSound(player, this, SoundEvents.AMETHYST_BLOCK_CHIME, SoundSource.NEUTRAL, 2.0F, 1.0F);
-            this.removeInteractionItem(player, interactionItem);
-            return InteractionResult.SUCCESS;
+            //this.level().broadcastEntityEvent(this, EntityEvent.IN_LOVE_HEARTS);
+            //this.level().playSound(player, this, SoundEvents.AMETHYST_BLOCK_CHIME, SoundSource.NEUTRAL, 2.0F, 1.0F);
+            //this.removeInteractionItem(player, interactionItem);
+            //return InteractionResult.SUCCESS;
+            player.containerMenu.sendAllDataToRemote();
+            this.resendPossiblyDesyncedEntityData((ServerPlayer) player);
+            return InteractionResult.FAIL;
         }
 
         if (itemInHand.isEmpty() && !interactionItem.isEmpty()) {
-            ItemStack itemToGive = interactionItem.copyWithCount(1);
-            this.setItemInHand(InteractionHand.MAIN_HAND, itemToGive);
-            this.removeInteractionItem(player, interactionItem);
-            this.level().playSound(player, this, SoundEvents.ALLAY_ITEM_GIVEN, SoundSource.NEUTRAL, 2.0F, 1.0F);
-            this.getBrain().setMemory(MemoryModuleType.LIKED_PLAYER, player.getUUID());
-            return InteractionResult.SUCCESS;
+            //ItemStack itemToGive = interactionItem.copyWithCount(1);
+            //this.setItemInHand(InteractionHand.MAIN_HAND, itemToGive);
+            //this.removeInteractionItem(player, interactionItem);
+            //this.level().playSound(player, this, SoundEvents.ALLAY_ITEM_GIVEN, SoundSource.NEUTRAL, 2.0F, 1.0F);
+            //this.getBrain().setMemory(MemoryModuleType.LIKED_PLAYER, player.getUUID());
+            //return InteractionResult.SUCCESS;
+            player.containerMenu.sendAllDataToRemote();
+            this.resendPossiblyDesyncedEntityData((ServerPlayer) player);
+            return InteractionResult.FAIL;
         }
 
         if (!itemInHand.isEmpty() && hand == InteractionHand.MAIN_HAND && interactionItem.isEmpty()) {
-            this.setItemSlot(EquipmentSlot.MAINHAND, ItemStack.EMPTY);
-            this.level().playSound(player, this, SoundEvents.ALLAY_ITEM_TAKEN, SoundSource.NEUTRAL, 2.0F, 1.0F);
-            this.swing(InteractionHand.MAIN_HAND);
+            //this.setItemSlot(EquipmentSlot.MAINHAND, ItemStack.EMPTY);
+            //this.level().playSound(player, this, SoundEvents.ALLAY_ITEM_TAKEN, SoundSource.NEUTRAL, 2.0F, 1.0F);
+            //this.swing(InteractionHand.MAIN_HAND);
 
             for (ItemStack itemStack : this.getInventory().removeAllItems()) {
-                BehaviorUtils.throwItem(this, itemStack, this.position());
+                //BehaviorUtils.throwItem(this, itemStack, this.position());
             }
 
-            this.getBrain().eraseMemory(MemoryModuleType.LIKED_PLAYER);
-            player.addItem(itemInHand);
-            return InteractionResult.SUCCESS;
+            //this.getBrain().eraseMemory(MemoryModuleType.LIKED_PLAYER);
+            //player.addItem(itemInHand);
+            //return InteractionResult.SUCCESS;
+            player.containerMenu.sendAllDataToRemote();
+            this.resendPossiblyDesyncedEntityData((ServerPlayer) player);
+            return InteractionResult.FAIL;
         } else {
             return super.mobInteract(player, hand);
         }
     }
 
     public void setJukeboxPlaying(final BlockPos jukebox, final boolean isPlaying) {
-        if (isPlaying) {
-            if (!this.isDancing()) {
-                this.jukeboxPos = jukebox;
-                this.setDancing(true);
-            }
-        } else if (jukebox.equals(this.jukeboxPos) || this.jukeboxPos == null) {
-            this.jukeboxPos = null;
-            this.setDancing(false);
-        }
+//        if (isPlaying) {
+//            if (!this.isDancing()) {
+//                this.jukeboxPos = jukebox;
+//                this.setDancing(true);
+//            }
+//        } else if (jukebox.equals(this.jukeboxPos) || this.jukeboxPos == null) {
+//            this.jukeboxPos = null;
+//            this.setDancing(false);
+//        }
     }
 
     @Override
@@ -352,11 +362,12 @@ public class Allay extends PathfinderMob implements InventoryCarrier, VibrationS
 
     @Override
     public boolean wantsToPickUp(final ServerLevel level, final ItemStack itemStack) {
-        ItemStack itemInHand = this.getItemInHand(InteractionHand.MAIN_HAND);
-        return !itemInHand.isEmpty()
-            && level.getGameRules().get(GameRules.MOB_GRIEFING)
-            && this.inventory.canAddItem(itemStack)
-            && this.allayConsidersItemEqual(itemInHand, itemStack);
+//        ItemStack itemInHand = this.getItemInHand(InteractionHand.MAIN_HAND);
+//        return !itemInHand.isEmpty()
+//            && level.getGameRules().get(GameRules.MOB_GRIEFING)
+//            && this.inventory.canAddItem(itemStack)
+//            && this.allayConsidersItemEqual(itemInHand, itemStack);
+        return false; // Cartridge
     }
 
     private boolean allayConsidersItemEqual(final ItemStack item1, final ItemStack item2) {
@@ -371,7 +382,7 @@ public class Allay extends PathfinderMob implements InventoryCarrier, VibrationS
 
     @Override
     protected void pickUpItem(final ServerLevel level, final ItemEntity entity) {
-        InventoryCarrier.pickUpItem(level, this, this, entity);
+//        InventoryCarrier.pickUpItem(level, this, this, entity);
     }
 
     @Override
@@ -379,22 +390,22 @@ public class Allay extends PathfinderMob implements InventoryCarrier, VibrationS
         return !this.onGround();
     }
 
-    @Override
-    public void updateDynamicGameEventListener(final BiConsumer<DynamicGameEventListener<?>, ServerLevel> action) {
-        if (this.level() instanceof ServerLevel serverLevel) {
-            action.accept(this.dynamicVibrationListener, serverLevel);
-            action.accept(this.dynamicJukeboxListener, serverLevel);
-        }
-    }
+//    @Override
+//    public void updateDynamicGameEventListener(final BiConsumer<DynamicGameEventListener<?>, ServerLevel> action) {
+//        if (this.level() instanceof ServerLevel serverLevel) {
+//            action.accept(this.dynamicVibrationListener, serverLevel);
+//            action.accept(this.dynamicJukeboxListener, serverLevel);
+//        }
+//    }
 
     public boolean isDancing() {
         return this.entityData.get(DATA_DANCING);
     }
 
     public void setDancing(final boolean isDancing) {
-        if (!this.level().isClientSide() && this.isEffectiveAi() && (!isDancing || !this.isPanicking())) {
-            this.entityData.set(DATA_DANCING, isDancing);
-        }
+//        if (!this.level().isClientSide() && this.isEffectiveAi() && (!isDancing || !this.isPanicking())) {
+//            this.entityData.set(DATA_DANCING, isDancing);
+//        }
     }
 
     private boolean shouldStopDancing() {
@@ -443,7 +454,7 @@ public class Allay extends PathfinderMob implements InventoryCarrier, VibrationS
     protected void addAdditionalSaveData(final ValueOutput output) {
         super.addAdditionalSaveData(output);
         this.writeInventoryToTag(output);
-        output.store("listener", VibrationSystem.Data.CODEC, this.vibrationData);
+        //output.store("listener", VibrationSystem.Data.CODEC, this.vibrationData);
         output.putLong("DuplicationCooldown", this.duplicationCooldown);
     }
 
@@ -451,7 +462,7 @@ public class Allay extends PathfinderMob implements InventoryCarrier, VibrationS
     protected void readAdditionalSaveData(final ValueInput input) {
         super.readAdditionalSaveData(input);
         this.readInventoryFromTag(input);
-        this.vibrationData = input.read("listener", VibrationSystem.Data.CODEC).orElseGet(VibrationSystem.Data::new);
+        //this.vibrationData = input.read("listener", VibrationSystem.Data.CODEC).orElseGet(VibrationSystem.Data::new);
         this.setDuplicationCooldown(input.getLongOr("DuplicationCooldown", 0)); // Paper - Load as long
     }
 
@@ -472,16 +483,17 @@ public class Allay extends PathfinderMob implements InventoryCarrier, VibrationS
     }
 
     public @Nullable Allay duplicateAllay() { // CraftBukkit - return allay
-        Allay allay = EntityTypes.ALLAY.create(this.level(), EntitySpawnReason.BREEDING);
-        if (allay != null) {
-            allay.snapTo(this.position());
-            allay.setPersistenceRequired();
-            allay.resetDuplicationCooldown();
-            this.resetDuplicationCooldown();
-            this.level().addFreshEntity(allay, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DUPLICATION); // CraftBukkit - reason for duplicated allay
-        }
+//        Allay allay = EntityTypes.ALLAY.create(this.level(), EntitySpawnReason.BREEDING);
+//        if (allay != null) {
+//            allay.snapTo(this.position());
+//            allay.setPersistenceRequired();
+//            allay.resetDuplicationCooldown();
+//            this.resetDuplicationCooldown();
+//            this.level().addFreshEntity(allay, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DUPLICATION); // CraftBukkit - reason for duplicated allay
+//        }
 
-        return allay; // CraftBukkit
+//        return allay; // CraftBukkit
+        return null; // Cartridge
     }
 
     public void resetDuplicationCooldown() {
@@ -489,11 +501,12 @@ public class Allay extends PathfinderMob implements InventoryCarrier, VibrationS
     }
 
     public boolean canDuplicate() {
-        return this.entityData.get(DATA_CAN_DUPLICATE);
+        //return this.entityData.get(DATA_CAN_DUPLICATE);
+        return false;
     }
 
     private void removeInteractionItem(final Player player, final ItemStack interactionItem) {
-        interactionItem.consume(1, player);
+//        interactionItem.consume(1, player);
     }
 
     @Override
@@ -501,113 +514,113 @@ public class Allay extends PathfinderMob implements InventoryCarrier, VibrationS
         return new Vec3(0.0, this.getEyeHeight() * 0.6, this.getBbWidth() * 0.1);
     }
 
-    @Override
-    public void handleEntityEvent(final byte id) {
-        if (id == EntityEvent.IN_LOVE_HEARTS) {
-            for (int i = 0; i < 3; i++) {
-                this.spawnHeartParticle();
-            }
-        } else {
-            super.handleEntityEvent(id);
-        }
-    }
-
-    private void spawnHeartParticle() {
-        double xd = this.random.nextGaussian() * 0.02;
-        double yd = this.random.nextGaussian() * 0.02;
-        double zd = this.random.nextGaussian() * 0.02;
-        this.level().addParticle(ParticleTypes.HEART, this.getRandomX(1.0), this.getRandomY() + 0.5, this.getRandomZ(1.0), xd, yd, zd);
-    }
-
-    @Override
-    public VibrationSystem.Data getVibrationData() {
-        return this.vibrationData;
-    }
-
-    @Override
-    public VibrationSystem.User getVibrationUser() {
-        return this.vibrationUser;
-    }
-
-    private class JukeboxListener implements GameEventListener {
-        private final PositionSource listenerSource;
-        private final int listenerRadius;
-
-        public JukeboxListener(final PositionSource listenerSource, final int listenerRadius) {
-            this.listenerSource = listenerSource;
-            this.listenerRadius = listenerRadius;
-        }
-
-        @Override
-        public PositionSource getListenerSource() {
-            return this.listenerSource;
-        }
-
-        @Override
-        public int getListenerRadius() {
-            return this.listenerRadius;
-        }
-
-        @Override
-        public boolean handleGameEvent(final ServerLevel level, final Holder<GameEvent> event, final GameEvent.Context context, final Vec3 sourcePosition) {
-            if (event.is(GameEvent.JUKEBOX_PLAY)) {
-                Allay.this.setJukeboxPlaying(BlockPos.containing(sourcePosition), true);
-                return true;
-            } else if (event.is(GameEvent.JUKEBOX_STOP_PLAY)) {
-                Allay.this.setJukeboxPlaying(BlockPos.containing(sourcePosition), false);
-                return true;
-            } else {
-                return false;
-            }
-        }
-    }
-
-    private class VibrationUser implements VibrationSystem.User {
-        private static final int VIBRATION_EVENT_LISTENER_RANGE = 16;
-        private final PositionSource positionSource = new EntityPositionSource(Allay.this, Allay.this.getEyeHeight());
-
-        @Override
-        public int getListenerRadius() {
-            return 16;
-        }
-
-        @Override
-        public PositionSource getPositionSource() {
-            return this.positionSource;
-        }
-
-        @Override
-        public boolean canReceiveVibration(final ServerLevel level, final BlockPos pos, final Holder<GameEvent> event, final GameEvent.Context context) {
-            if (Allay.this.isNoAi()) {
-                return false;
-            }
-
-            Optional<GlobalPos> maybeGlobalPos = Allay.this.getBrain().getMemory(MemoryModuleType.LIKED_NOTEBLOCK_POSITION);
-            if (maybeGlobalPos.isEmpty()) {
-                return true;
-            }
-
-            GlobalPos globalPos = maybeGlobalPos.get();
-            return globalPos.isCloseEnough(level.dimension(), Allay.this.blockPosition(), 1024) && globalPos.pos().equals(pos);
-        }
-
-        @Override
-        public void onReceiveVibration(
-            final ServerLevel level,
-            final BlockPos pos,
-            final Holder<GameEvent> event,
-            final @Nullable Entity sourceEntity,
-            final @Nullable Entity projectileOwner,
-            final float receivingDistance
-        ) {
-            if (event.is(GameEvent.NOTE_BLOCK_PLAY)) {
-                AllayAi.hearNoteblock(Allay.this, new BlockPos(pos));
-            }
-        }
-
-        @Override
-        public TagKey<GameEvent> getListenableEvents() {
-            return GameEventTags.ALLAY_CAN_LISTEN;
-        }
-    }
+//    @Override
+//    public void handleEntityEvent(final byte id) {
+//        if (id == EntityEvent.IN_LOVE_HEARTS) {
+//            for (int i = 0; i < 3; i++) {
+//                this.spawnHeartParticle();
+//            }
+//        } else {
+//            super.handleEntityEvent(id);
+//        }
+//    }
+
+//    private void spawnHeartParticle() {
+//        double xd = this.random.nextGaussian() * 0.02;
+//        double yd = this.random.nextGaussian() * 0.02;
+//        double zd = this.random.nextGaussian() * 0.02;
+//        this.level().addParticle(ParticleTypes.HEART, this.getRandomX(1.0), this.getRandomY() + 0.5, this.getRandomZ(1.0), xd, yd, zd);
+//    }
+
+//    @Override
+//    public VibrationSystem.Data getVibrationData() {
+//        return this.vibrationData;
+//    }
+
+//    @Override
+//    public VibrationSystem.User getVibrationUser() {
+//        return this.vibrationUser;
+//    }
+
+//    private class JukeboxListener implements GameEventListener {
+//        private final PositionSource listenerSource;
+//        private final int listenerRadius;
+
+//        public JukeboxListener(final PositionSource listenerSource, final int listenerRadius) {
+//            this.listenerSource = listenerSource;
+//            this.listenerRadius = listenerRadius;
+//        }
+
+//        @Override
+//        public PositionSource getListenerSource() {
+//            return this.listenerSource;
+//        }
+
+//        @Override
+//        public int getListenerRadius() {
+//            return this.listenerRadius;
+//        }
+
+//        @Override
+//        public boolean handleGameEvent(final ServerLevel level, final Holder<GameEvent> event, final GameEvent.Context context, final Vec3 sourcePosition) {
+//            if (event.is(GameEvent.JUKEBOX_PLAY)) {
+//                Allay.this.setJukeboxPlaying(BlockPos.containing(sourcePosition), true);
+//                return true;
+//            } else if (event.is(GameEvent.JUKEBOX_STOP_PLAY)) {
+//                Allay.this.setJukeboxPlaying(BlockPos.containing(sourcePosition), false);
+//                return true;
+//            } else {
+//                return false;
+//            }
+//        }
+//    }
+
+//    private class VibrationUser implements VibrationSystem.User {
+//        private static final int VIBRATION_EVENT_LISTENER_RANGE = 16;
+//        private final PositionSource positionSource = new EntityPositionSource(Allay.this, Allay.this.getEyeHeight());
+
+//        @Override
+//        public int getListenerRadius() {
+//            return 16;
+//        }
+
+//        @Override
+//        public PositionSource getPositionSource() {
+//            return this.positionSource;
+//        }
+
+//        @Override
+//        public boolean canReceiveVibration(final ServerLevel level, final BlockPos pos, final Holder<GameEvent> event, final GameEvent.Context context) {
+//            if (Allay.this.isNoAi()) {
+//                return false;
+//            }
+
+//            Optional<GlobalPos> maybeGlobalPos = Allay.this.getBrain().getMemory(MemoryModuleType.LIKED_NOTEBLOCK_POSITION);
+//            if (maybeGlobalPos.isEmpty()) {
+//                return true;
+//            }
+
+//            GlobalPos globalPos = maybeGlobalPos.get();
+//            return globalPos.isCloseEnough(level.dimension(), Allay.this.blockPosition(), 1024) && globalPos.pos().equals(pos);
+//        }
+
+//        @Override
+//        public void onReceiveVibration(
+//            final ServerLevel level,
+//            final BlockPos pos,
+//            final Holder<GameEvent> event,
+//            final @Nullable Entity sourceEntity,
+//            final @Nullable Entity projectileOwner,
+//            final float receivingDistance
+//        ) {
+//            if (event.is(GameEvent.NOTE_BLOCK_PLAY)) {
+//                AllayAi.hearNoteblock(Allay.this, new BlockPos(pos));
+//            }
+//        }
+
+//        @Override
+//        public TagKey<GameEvent> getListenableEvents() {
+//            return GameEventTags.ALLAY_CAN_LISTEN;
+//        }
+//    }
 }
