From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Boy0000 <sivertpaulsen2@gmail.com>
Date: Sun, 4 Jan 2026 21:32:18 +0100
Subject: [PATCH] Branding


diff --git a/src/main/java/io/papermc/paper/ServerBuildInfo.java b/src/main/java/io/papermc/paper/ServerBuildInfo.java
index 652ff54e7c50412503725d628bfe72ed03059790..38283a1ea0b9f5ae4a9b28c0836b7ca531cfd51c 100644
--- a/src/main/java/io/papermc/paper/ServerBuildInfo.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfo.java
@@ -19,6 +19,8 @@ public interface ServerBuildInfo {
      */
     Key BRAND_PAPER_ID = Key.key("papermc", "paper");
 
+    Key BRAND_CARTRIDGE_ID = Key.key("mineinabyss", "cartridge");
+
     /**
      * Gets the {@code ServerBuildInfo}.
      *
