From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Boy0000 <sivertpaulsen2@gmail.com>
Date: Sun, 4 Jan 2026 21:32:18 +0100
Subject: [PATCH] Branding


diff --git a/src/main/java/io/papermc/paper/ServerBuildInfo.java b/src/main/java/io/papermc/paper/ServerBuildInfo.java
index 4e099b584c3435ba1ebce492011d35ae1ac0cdf8..741e5c8123e5c20948f486d594d3d4713bca9422 100644
--- a/src/main/java/io/papermc/paper/ServerBuildInfo.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfo.java
@@ -17,6 +17,8 @@ public interface ServerBuildInfo {
      */
     Key BRAND_PAPER_ID = Key.key("papermc", "paper");
 
+    Key BRAND_CARTRIDGE_ID = Key.key("mineinabyss", "cartridge");
+
     /**
      * Gets the {@code ServerBuildInfo}.
      *
