From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Boy <sivertpaulsen2@gmail.com>
Date: Tue, 18 Jun 2024 19:55:50 +0200
Subject: [PATCH] Datacomponent fixes


diff --git a/src/main/java/io/papermc/paper/component/DataComponentTypes.java b/src/main/java/io/papermc/paper/component/DataComponentTypes.java
index b404d32baa750f0a5d27a9a3fa1b454d1bd4e4f6..f5cc964e0976b5c256cc2271c31942cc87a34388 100644
--- a/src/main/java/io/papermc/paper/component/DataComponentTypes.java
+++ b/src/main/java/io/papermc/paper/component/DataComponentTypes.java
@@ -86,7 +86,7 @@ public final class DataComponentTypes {
     // block_entity_data
     public static final DataComponentType.Valued<MusicInstrument> INSTRUMENT = valued("instrument");
     public static final DataComponentType.Valued<Integer> OMINOUS_BOTTLE_AMPLIFIER = valued("ominous_bottle_amplifier");
-    public static final DataComponentType.Valued<JukeboxPlayable> JUKEBOX_PLAYALE = valued("jukebox_playable");
+    public static final DataComponentType.Valued<JukeboxPlayable> JUKEBOX_PLAYABLE = valued("jukebox_playable");
     public static final DataComponentType.Valued<List<Key>> RECIPES = valued("recipes");
     public static final DataComponentType.Valued<LodestoneTracker> LODESTONE_TRACKER = valued("lodestone_tracker");
     public static final DataComponentType.Valued<FireworkEffect> FIREWORK_EXPLOSION = valued("firework_explosion");
diff --git a/src/main/java/io/papermc/paper/component/item/ComponentTypesBridge.java b/src/main/java/io/papermc/paper/component/item/ComponentTypesBridge.java
index f05ccb7973028c9853de57406a514113b4257079..59db78f559404777ba6e711a0b2c267d750ec6c8 100644
--- a/src/main/java/io/papermc/paper/component/item/ComponentTypesBridge.java
+++ b/src/main/java/io/papermc/paper/component/item/ComponentTypesBridge.java
@@ -47,7 +47,7 @@ interface ComponentTypesBridge {
 
     SuspiciousStewEffects.Builder suspiciousStewEffects();
 
-    BundleContents.Builder mapItemColor();
+    MapItemColor.Builder mapItemColor();
 
     MapID.Builder mapId();
 
diff --git a/src/main/java/io/papermc/paper/component/item/MapItemColor.java b/src/main/java/io/papermc/paper/component/item/MapItemColor.java
index 4c606e1d4856a912dc407b6047cb9e024e822cb3..025e2edd791db72b87f4e43bdda6647cfe6b3d7e 100644
--- a/src/main/java/io/papermc/paper/component/item/MapItemColor.java
+++ b/src/main/java/io/papermc/paper/component/item/MapItemColor.java
@@ -9,7 +9,7 @@ import org.jetbrains.annotations.Contract;
 public interface MapItemColor {
 
     @Contract(value = "-> new", pure = true)
-    static BundleContents.@NonNull Builder mapItemColor() {
+    static MapItemColor.@NonNull Builder mapItemColor() {
         return ComponentTypesBridge.bridge().mapItemColor();
     }
 
