From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Boy <sivertpaulsen2@gmail.com>
Date: Sat, 15 Jun 2024 01:36:28 +0200
Subject: [PATCH] Block & Falldamage tweaks


diff --git a/src/main/java/net/minecraft/data/tags/VanillaBlockTagsProvider.java b/src/main/java/net/minecraft/data/tags/VanillaBlockTagsProvider.java
index 118083ab8648ecc620c92a473044a42dfaeb0021..cb59947cc29f7943010977631a47f4cf9fb28de2 100644
--- a/src/main/java/net/minecraft/data/tags/VanillaBlockTagsProvider.java
+++ b/src/main/java/net/minecraft/data/tags/VanillaBlockTagsProvider.java
@@ -747,7 +747,7 @@ public class VanillaBlockTagsProvider extends IntrinsicHolderTagsProvider<Block>
                 Blocks.CAVE_VINES,
                 Blocks.CAVE_VINES_PLANT
             );
-        this.tag(BlockTags.FALL_DAMAGE_RESETTING).addTag(BlockTags.CLIMBABLE).add(Blocks.SWEET_BERRY_BUSH, Blocks.COBWEB);
+        this.tag(BlockTags.FALL_DAMAGE_RESETTING);
         this.tag(BlockTags.PIGLIN_REPELLENTS)
             .add(Blocks.SOUL_FIRE)
             .add(Blocks.SOUL_TORCH)
diff --git a/src/main/java/net/minecraft/network/protocol/common/ClientboundUpdateTagsPacket.java b/src/main/java/net/minecraft/network/protocol/common/ClientboundUpdateTagsPacket.java
index 0e0c9a54df98413b0ad4d195dd5ba21132cf4065..027d4e5dffb0d7f085b9a7792ad61b90cb37d5c2 100644
--- a/src/main/java/net/minecraft/network/protocol/common/ClientboundUpdateTagsPacket.java
+++ b/src/main/java/net/minecraft/network/protocol/common/ClientboundUpdateTagsPacket.java
@@ -1,13 +1,27 @@
 package net.minecraft.network.protocol.common;
 
+import java.util.ArrayList;
+import java.util.List;
 import java.util.Map;
+import java.util.Optional;
+
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+import it.unimi.dsi.fastutil.ints.IntList;
 import net.minecraft.core.Registry;
+import net.minecraft.core.registries.BuiltInRegistries;
+import net.minecraft.core.registries.Registries;
+import net.minecraft.nbt.Tag;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.codec.StreamCodec;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.network.protocol.PacketType;
 import net.minecraft.resources.ResourceKey;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.tags.BlockTags;
 import net.minecraft.tags.TagNetworkSerialization;
+import net.minecraft.world.level.block.Block;
+import org.jetbrains.annotations.Nullable;
 
 public class ClientboundUpdateTagsPacket implements Packet<ClientCommonPacketListener> {
     public static final StreamCodec<FriendlyByteBuf, ClientboundUpdateTagsPacket> STREAM_CODEC = Packet.codec(
@@ -37,6 +51,26 @@ public class ClientboundUpdateTagsPacket implements Packet<ClientCommonPacketLis
         listener.handleUpdateTags(this);
     }
 
+    // Cartridge start
+    @Override
+    public void onPacketDispatch(@Nullable ServerPlayer player) {
+        // if null aka not play-phase return
+        if (player == null) {
+            return;
+        }
+        Map<ResourceLocation, IntList> regTags = Optional.ofNullable(tags.get(BuiltInRegistries.BLOCK.key())).map(n -> n.tags).orElse(null);
+        // If the packet does not contain the block-registry, return as tag was not modified
+        if (regTags == null) {
+            return;
+        }
+        IntList climbableIds = regTags.get(BlockTags.CLIMBABLE.location());
+        // If climbable is not in the packet, return
+        if (climbableIds == null) return;
+        // Set the client-side climbable state based on packet
+        player.hasClientSideClimbable = !climbableIds.isEmpty();
+    }
+    // Cartridge end
+
     public Map<ResourceKey<? extends Registry<?>>, TagNetworkSerialization.NetworkPayload> getTags() {
         return this.tags;
     }
diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 28723ab8bd72ff0db5e21c68f44f0a8d96b37653..ab958711603f50904c70a2d4e206e41ae2e43c6c 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -339,6 +339,10 @@ public class ServerPlayer extends net.minecraft.world.entity.player.Player imple
     }
     // Paper end - rewrite chunk system
 
+    // Cartridge start
+    public boolean hasClientSideClimbable = true;
+    // Cartridge end
+
     public ServerPlayer(MinecraftServer server, ServerLevel world, GameProfile profile, ClientInformation clientOptions) {
         super(world, world.getSharedSpawnPos(), world.getSharedSpawnAngle(), profile);
         this.chatVisibility = ChatVisiblity.FULL;
@@ -1199,9 +1203,11 @@ public class ServerPlayer extends net.minecraft.world.entity.player.Player imple
                 if (getRootVehicle() instanceof net.minecraft.world.entity.vehicle.AbstractMinecart && level().purpurConfig.minecartControllable && !level().purpurConfig.minecartControllableFallDamage) {
                     return false;
                 }
-                if (getRootVehicle() instanceof net.minecraft.world.entity.vehicle.Boat && !level().purpurConfig.boatsDoFallDamage) {
-                    return false;
-                }
+                // Cartridge start - boats dont do falldamage
+                //if (getRootVehicle() instanceof net.minecraft.world.entity.vehicle.Boat && !level().purpurConfig.boatsDoFallDamage) {
+                //    return false;
+                //}
+                // Cartridge end
             }
             // Purpur end
             boolean flag = this.server.isDedicatedServer() && this.isPvpAllowed() && source.is(DamageTypeTags.IS_FALL);
@@ -1690,6 +1696,17 @@ public class ServerPlayer extends net.minecraft.world.entity.player.Player imple
                 this.spawnExtraParticlesOnFall = false;
             }
 
+            if (hasLandedInLiquid()) {
+                if (fallDistance < getMaxFallDistance() && level().getBlockStates(this.getBoundingBox().move(0,-2,0)).allMatch(b -> !b.getFluidState().isEmpty() || b.isSolid())) onGround = true;
+                else if (fallDistance < getMaxFallDistance() || !iblockdata.getFluidState().isEmpty()) onGround = false;
+                if (onGround && !iblockdata.getFluidState().isEmpty() && (isEyeInFluid(FluidTags.WATER))) onGround = false;
+                if (iblockdata.isSolid() && iblockdata.getFluidState().isEmpty()) onGround = true;
+
+            } else if (fallDistance >= getMaxFallDistance() && !iblockdata.getFluidState().isEmpty()) {
+                if (this.getDeltaMovement().y() > 9.999999747378752E-6D) resetFallDistance();
+                onGround = true;
+            }
+
             super.checkFallDamage(yDifference, onGround, iblockdata, blockposition);
         }
     }
@@ -1699,7 +1716,8 @@ public class ServerPlayer extends net.minecraft.world.entity.player.Player imple
         super.onExplosionHit(entity);
         this.currentImpulseImpactPos = this.position();
         this.currentExplosionCause = entity;
-        this.setIgnoreFallDamageFromCurrentImpulse(entity != null && entity.getType() == EntityType.WIND_CHARGE);
+        //this.setIgnoreFallDamageFromCurrentImpulse(entity != null && entity.getType() == EntityType.WIND_CHARGE); // Cartridge
+        this.setIgnoreFallDamageFromCurrentImpulse(false); // Cartridge
     }
 
     @Override
diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index e0962547709d6951cc98da94f028bd4d5b7b25dd..245d5fd3a092f4727bf32983eb0a6871d17a0df6 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -1,5 +1,6 @@
 package net.minecraft.server.network;
 
+import com.destroystokyo.paper.utils.PaperPluginLogger;
 import com.google.common.collect.Lists;
 import com.google.common.primitives.Floats;
 import com.mojang.authlib.GameProfile;
@@ -7,6 +8,7 @@ import com.mojang.brigadier.ParseResults;
 import com.mojang.brigadier.StringReader;
 import com.mojang.brigadier.suggestion.Suggestions;
 import com.mojang.logging.LogUtils;
+import io.papermc.paper.adventure.providers.ComponentLoggerProviderImpl;
 import it.unimi.dsi.fastutil.ints.Int2ObjectMap.Entry;
 import it.unimi.dsi.fastutil.ints.Int2ObjectMaps;
 import it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap;
@@ -135,6 +137,7 @@ import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.stats.ServerRecipeBook;
+import net.minecraft.tags.BlockTags;
 import net.minecraft.util.FutureChain;
 import net.minecraft.util.Mth;
 import net.minecraft.util.SignatureValidator;
@@ -600,9 +603,13 @@ public class ServerGamePacketListenerImpl extends ServerCommonPacketListenerImpl
                 if (entity instanceof LivingEntity) {
                     LivingEntity entityliving = (LivingEntity) entity;
 
-                    if (entityliving.onClimbable()) {
-                        entityliving.resetFallDistance();
+                    // Cartridge start
+                    if (entityliving.getInBlockState().is(BlockTags.CLIMBABLE) && (!(entityliving instanceof ServerPlayer serverPlayer) || serverPlayer.hasClientSideClimbable)) {
+                        // If on climbable and the last climbable is not empty, they are currently climbing
+                        // If it is not present, they landed on the ladder, and we want to retain their falldistance
+                        if (entityliving.getLastClimbablePos().isPresent()) entityliving.resetFallDistance();
                     }
+                    // Cartridge end
                 }
 
                 entity.move(MoverType.PLAYER, new Vec3(d6, d7, d8));
@@ -714,6 +721,7 @@ public class ServerGamePacketListenerImpl extends ServerCommonPacketListenerImpl
                 this.vehicleLastGoodX = entity.getX();
                 this.vehicleLastGoodY = entity.getY();
                 this.vehicleLastGoodZ = entity.getZ();
+                this.player.fallDistance = entity.fallDistance; // Cartridge
             }
 
         }
@@ -1577,7 +1585,7 @@ public class ServerGamePacketListenerImpl extends ServerCommonPacketListenerImpl
                             if (teleportBack) {
                             // Paper end - Add fail move event
                                 this.internalTeleport(d3, d4, d5, f, f1, Collections.emptySet()); // CraftBukkit - SPIGOT-1807: Don't call teleport event, when the client thinks the player is falling, because the chunks are not loaded on the client yet.
-                                this.player.doCheckFallDamage(this.player.getX() - d3, this.player.getY() - d4, this.player.getZ() - d5, packet.isOnGround());
+                                this.player.doCheckFallDamage(this.player.getX() - d3, this.player.getY() - d4, this.player.getZ() - d5, packet.isOnGround() || this.player.isInWater()); // Cartridge - Make water not negate falldamage
                             } else {
                                 // CraftBukkit start - fire PlayerMoveEvent
                                 // Reset to old location first
@@ -1658,7 +1666,7 @@ public class ServerGamePacketListenerImpl extends ServerCommonPacketListenerImpl
                                 this.player.doCheckFallDamage(this.player.getX() - d3, this.player.getY() - d4, this.player.getZ() - d5, packet.isOnGround());
                                 this.player.setKnownMovement(vec3d);
                                 if (flag1) {
-                                    this.player.resetFallDistance();
+                                    //this.player.resetFallDistance();
                                 }
 
                                 if (packet.isOnGround() || this.player.hasLandedInLiquid() || this.player.onClimbable() || this.player.isSpectator() || flag || flag4) {
diff --git a/src/main/java/net/minecraft/tags/TagNetworkSerialization.java b/src/main/java/net/minecraft/tags/TagNetworkSerialization.java
index b6c1c9d6bef3dd234c9bddb628ac6620ad12b854..d4a8b1986cbd5191b9b3a8813278c9ff01df82f9 100644
--- a/src/main/java/net/minecraft/tags/TagNetworkSerialization.java
+++ b/src/main/java/net/minecraft/tags/TagNetworkSerialization.java
@@ -47,7 +47,7 @@ public class TagNetworkSerialization {
         return new TagNetworkSerialization.NetworkPayload(map);
     }
 
-    static <T> void deserializeTagsFromNetwork(
+    public static <T> void deserializeTagsFromNetwork(
         ResourceKey<? extends Registry<T>> registryKey,
         Registry<T> registry,
         TagNetworkSerialization.NetworkPayload serialized,
@@ -61,7 +61,7 @@ public class TagNetworkSerialization {
     }
 
     public static final class NetworkPayload {
-        final Map<ResourceLocation, IntList> tags;
+        public final Map<ResourceLocation, IntList> tags;
 
         NetworkPayload(Map<ResourceLocation, IntList> contents) {
             this.tags = contents;
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 3c467580195355fb1a04568e628f601365fdbfe0..0618d746cdc90ec2012c0c5ea790eeb168a70856 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -6,6 +6,7 @@ import com.google.common.collect.Lists;
 import com.google.common.collect.Sets;
 import com.google.common.collect.UnmodifiableIterator;
 import com.mojang.logging.LogUtils;
+import io.papermc.paper.adventure.PaperAdventure;
 import it.unimi.dsi.fastutil.doubles.DoubleList;
 import it.unimi.dsi.fastutil.doubles.DoubleListIterator;
 import it.unimi.dsi.fastutil.floats.FloatArraySet;
@@ -43,6 +44,7 @@ import net.minecraft.core.RegistryAccess;
 import net.minecraft.core.SectionPos;
 import net.minecraft.core.particles.BlockParticleOption;
 import net.minecraft.core.particles.ParticleTypes;
+import net.minecraft.core.registries.BuiltInRegistries;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.nbt.DoubleTag;
 import net.minecraft.nbt.FloatTag;
@@ -83,6 +85,7 @@ import net.minecraft.world.InteractionResult;
 import net.minecraft.world.Nameable;
 import net.minecraft.world.damagesource.DamageSource;
 import net.minecraft.world.damagesource.DamageSources;
+import net.minecraft.world.entity.ai.attributes.Attributes;
 import net.minecraft.world.entity.item.ItemEntity;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.entity.projectile.Projectile;
@@ -954,7 +957,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
 
         if (this.isInLava()) {
             this.lavaHurt();
-            this.fallDistance *= 0.5F;
+            //this.fallDistance *= 0.5F; // Cartridge
             // CraftBukkit start
         } else {
             this.lastLavaContact = null;
@@ -1217,7 +1220,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                     BlockHitResult movingobjectpositionblock = this.level().clip(new ClipContext(this.position(), this.position().add(vec3d1), ClipContext.Block.FALLDAMAGE_RESETTING, ClipContext.Fluid.WATER, this));
 
                     if (movingobjectpositionblock.getType() != HitResult.Type.MISS) {
-                        this.resetFallDistance();
+                        //this.resetFallDistance(); // Cartridge - Make water not negate falldamage
                     }
                 }
 
@@ -1434,7 +1437,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
     // Moonrise end - Avoid streams for block retrieval in Entity#move
 
     private boolean isStateClimbable(BlockState state) {
-        return state.is(BlockTags.CLIMBABLE) || state.is(Blocks.POWDER_SNOW);
+        return (state.is(BlockTags.CLIMBABLE) && (!(this instanceof ServerPlayer serverPlayer) || serverPlayer.hasClientSideClimbable)) || state.is(Blocks.POWDER_SNOW);
     }
 
     private boolean vibrationAndSoundEffectsFromBlock(BlockPos pos, BlockState state, boolean playSound, boolean emitEvent, Vec3 movement) {
@@ -2044,7 +2047,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
     }
 
     protected void checkFallDamage(double heightDifference, boolean onGround, BlockState state, BlockPos landedPosition) {
-        if (onGround) {
+        if (onGround || (state.is(BlockTags.CLIMBABLE) && this instanceof ServerPlayer serverPlayer && serverPlayer.hasClientSideClimbable)) {
             if (this.fallDistance > 0.0F) {
                 state.getBlock().fallOn(this.level(), state, landedPosition, this, this.fallDistance);
                 this.level().gameEvent((Holder) GameEvent.HIT_GROUND, this.position, GameEvent.Context.of(this, (BlockState) this.mainSupportingBlockPos.map((blockposition1) -> {
@@ -2077,7 +2080,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                 }
             }
 
-            return false;
+            return true; // Cartridge
         }
     }
 
@@ -2148,7 +2151,10 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                 this.doWaterSplashEffect();
             }
 
-            this.resetFallDistance();
+            // Cartridge start - Make water not negate falldamage
+            //this.resetFallDistance();
+            if (!(this instanceof ServerPlayer)) this.resetFallDistance();
+            // Cartridge end
             this.wasTouchingWater = true;
             this.clearFire();
         } else {
@@ -3301,6 +3307,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                     this.stopRiding();
                 }
 
+                this.causeFallDamage(this.fallDistance, 1.0f, this.damageSources().fall());
                 this.setPose(net.minecraft.world.entity.Pose.STANDING);
                 this.vehicle = entity;
                 this.vehicle.addPassenger(this);
@@ -3314,9 +3321,12 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
         }
     }
 
+    // Cartridge start - Prevent entering vehicle when falling
     protected boolean canRide(Entity entity) {
-        return !this.isShiftKeyDown() && this.boardingCooldown <= 0;
+        boolean isSafeFallDistance = (!(this instanceof net.minecraft.world.entity.LivingEntity livingEntity) || this.fallDistance < livingEntity.getAttributeValue(Attributes.SAFE_FALL_DISTANCE));
+        return isSafeFallDistance && !this.isShiftKeyDown() && this.boardingCooldown <= 0;
     }
+    // Cartridge end - Prevent entering vehicle when falling
 
     public void ejectPassengers() {
         for (int i = this.passengers.size() - 1; i >= 0; --i) {
@@ -3884,7 +3894,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
     }
 
     public void makeStuckInBlock(BlockState state, Vec3 multiplier) {
-        this.resetFallDistance();
+        //this.resetFallDistance(); // Cartridge
         this.stuckSpeedMultiplier = multiplier;
     }
 
diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 47e031130ca30cbe3bb9917cb9a43612ec73b538..3ffa15668780edb4412f01946d0fb9af372e01df 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -1,12 +1,9 @@
 package net.minecraft.world.entity;
 
 import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Function;
 import com.google.common.base.Objects;
-import com.google.common.collect.ImmutableList;
-import com.google.common.collect.ImmutableMap;
-import com.google.common.collect.Iterables;
-import com.google.common.collect.Lists;
-import com.google.common.collect.Maps;
+import com.google.common.collect.*;
 import com.mojang.datafixers.util.Pair;
 import com.mojang.logging.LogUtils;
 import com.mojang.serialization.DataResult;
@@ -14,17 +11,6 @@ import com.mojang.serialization.Dynamic;
 import it.unimi.dsi.fastutil.doubles.DoubleDoubleImmutablePair;
 import it.unimi.dsi.fastutil.objects.Reference2ObjectArrayMap;
 import it.unimi.dsi.fastutil.objects.Reference2ObjectMap;
-import java.util.Collection;
-import java.util.ConcurrentModificationException;
-import java.util.Iterator;
-import java.util.List;
-import java.util.Map;
-import java.util.Map.Entry;
-import java.util.Optional;
-import java.util.Set;
-import java.util.function.Predicate;
-import javax.annotation.Nonnull;
-import javax.annotation.Nullable;
 import net.minecraft.BlockUtil;
 import net.minecraft.Util;
 import net.minecraft.advancements.CriteriaTriggers;
@@ -38,19 +24,8 @@ import net.minecraft.core.particles.BlockParticleOption;
 import net.minecraft.core.particles.ItemParticleOption;
 import net.minecraft.core.particles.ParticleOptions;
 import net.minecraft.core.particles.ParticleTypes;
-import net.minecraft.nbt.CompoundTag;
-import net.minecraft.nbt.FloatTag;
-import net.minecraft.nbt.IntTag;
-import net.minecraft.nbt.ListTag;
-import net.minecraft.nbt.NbtOps;
-import net.minecraft.nbt.Tag;
-import net.minecraft.network.protocol.game.ClientboundAddEntityPacket;
-import net.minecraft.network.protocol.game.ClientboundAnimatePacket;
-import net.minecraft.network.protocol.game.ClientboundEntityEventPacket;
-import net.minecraft.network.protocol.game.ClientboundRemoveMobEffectPacket;
-import net.minecraft.network.protocol.game.ClientboundSetEquipmentPacket;
-import net.minecraft.network.protocol.game.ClientboundTakeItemEntityPacket;
-import net.minecraft.network.protocol.game.ClientboundUpdateMobEffectPacket;
+import net.minecraft.nbt.*;
+import net.minecraft.network.protocol.game.*;
 import net.minecraft.network.syncher.EntityDataAccessor;
 import net.minecraft.network.syncher.EntityDataSerializers;
 import net.minecraft.network.syncher.SynchedEntityData;
@@ -63,12 +38,7 @@ import net.minecraft.sounds.SoundEvent;
 import net.minecraft.sounds.SoundEvents;
 import net.minecraft.sounds.SoundSource;
 import net.minecraft.stats.Stats;
-import net.minecraft.tags.BlockTags;
-import net.minecraft.tags.DamageTypeTags;
-import net.minecraft.tags.EntityTypeTags;
-import net.minecraft.tags.FluidTags;
-import net.minecraft.tags.ItemTags;
-import net.minecraft.tags.TagKey;
+import net.minecraft.tags.*;
 import net.minecraft.util.Mth;
 import net.minecraft.world.Difficulty;
 import net.minecraft.world.InteractionHand;
@@ -132,14 +102,14 @@ import net.minecraft.world.phys.HitResult;
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.scores.PlayerTeam;
 import net.minecraft.world.scores.Scoreboard;
+import org.jetbrains.annotations.Nullable;
 import org.slf4j.Logger;
 
 // CraftBukkit start
-import java.util.ArrayList;
-import java.util.HashSet;
-import java.util.Set;
+import java.util.*;
+import java.util.function.Predicate;
+
 import com.google.common.base.Function;
-import java.util.UUID;
 import org.bukkit.Location;
 import org.bukkit.craftbukkit.attribute.CraftAttributeMap;
 import org.bukkit.craftbukkit.event.CraftEventFactory;
@@ -155,6 +125,8 @@ import org.bukkit.event.entity.EntityRemoveEvent;
 import org.bukkit.event.entity.EntityResurrectEvent;
 import org.bukkit.event.entity.EntityTeleportEvent;
 import org.bukkit.event.player.PlayerItemConsumeEvent;
+
+import javax.annotation.Nonnull;
 // CraftBukkit end
 
 public abstract class LivingEntity extends Entity implements Attackable, net.caffeinemc.mods.lithium.common.util.change_tracking.ChangeSubscriber.CountChangeSubscriber<ItemStack>, net.caffeinemc.mods.lithium.common.entity.EquipmentEntity, net.caffeinemc.mods.lithium.common.entity.EquipmentEntity.TickableEnchantmentTrackingEntity, net.caffeinemc.mods.lithium.common.entity.EquipmentEntity.EquipmentTrackingEntity { // Leaf - Lithium equipment tracking
@@ -384,7 +356,7 @@ public abstract class LivingEntity extends Entity implements Attackable, net.caf
         Level world = this.level();
 
         if (world instanceof ServerLevel worldserver) {
-            if (onGround && this.fallDistance > 0.0F) {
+            if ((onGround || isInWater()) && this.fallDistance > 0.0F) { // Cartridge - Make water not negate falldamage
                 this.onChangedBlock(worldserver, landedPosition);
                 double d1 = this.getAttributeValue(Attributes.SAFE_FALL_DISTANCE);
 
@@ -407,21 +379,29 @@ public abstract class LivingEntity extends Entity implements Attackable, net.caf
                     double d8 = Math.min((double) (0.2F + f / 15.0F), 2.5D);
                     int i = (int) (150.0D * d8);
 
+                    // Cartridge start - only show block-particles when on ground
                     // CraftBukkit start - visiblity api
-                    if (this instanceof ServerPlayer) {
-                        ((ServerLevel) this.level()).sendParticles((ServerPlayer) this, new BlockParticleOption(ParticleTypes.BLOCK, state), d2, d3, d4, i, 0.0D, 0.0D, 0.0D, 0.15000000596046448D, false);
-                    } else {
-                        ((ServerLevel) this.level()).sendParticles(new BlockParticleOption(ParticleTypes.BLOCK, state), d2, d3, d4, i, 0.0D, 0.0D, 0.0D, 0.15000000596046448D);
+                    if (onGround) {
+                        if (this instanceof ServerPlayer) {
+                            ((ServerLevel) this.level()).sendParticles((ServerPlayer) this, new BlockParticleOption(ParticleTypes.BLOCK, state), d2, d3, d4, i, 0.0D, 0.0D, 0.0D, 0.15000000596046448D, false);
+                        } else {
+                            ((ServerLevel) this.level()).sendParticles(new BlockParticleOption(ParticleTypes.BLOCK, state), d2, d3, d4, i, 0.0D, 0.0D, 0.0D, 0.15000000596046448D);
+                        }
                     }
                     // CraftBukkit end
+                    // Cartridge end
                 }
             }
         }
 
-        super.checkFallDamage(heightDifference, onGround, state, landedPosition);
+        // Cartridge start
+        boolean silentOnClimbable = this.getInBlockState().is(BlockTags.CLIMBABLE) && getLastClimbablePos().isPresent();
+        super.checkFallDamage(heightDifference, onGround || (silentOnClimbable && getLastClimbablePos().isEmpty()), state, landedPosition);
+
         if (onGround) {
             this.lastClimbablePos = Optional.empty();
         }
+        // Cartridge end
 
     }
 
@@ -2168,6 +2148,11 @@ public abstract class LivingEntity extends Entity implements Attackable, net.caf
     // CraftBukkit end
 
     public Optional<BlockPos> getLastClimbablePos() {
+        // Cartridge start - correctly set lastClimbablePos
+        if (this instanceof ServerPlayer serverPlayer && !serverPlayer.hasClientSideClimbable) {
+            this.lastClimbablePos = Optional.empty();
+        }
+        // Cartridge end - correctly set lastClimbablePos
         return this.lastClimbablePos;
     }
 
@@ -2200,7 +2185,9 @@ public abstract class LivingEntity extends Entity implements Attackable, net.caf
 
     public Boolean onClimbable(boolean loadChunk) {
         // Gale end - don't load chunks to activate climbing entities
-        if (this.isSpectator()) {
+        if (this instanceof ServerPlayer serverPlayer && !serverPlayer.hasClientSideClimbable) { // Cartridge
+            return false;
+        } else if (this.isSpectator()) {
             return false;
         } else {
             BlockPos blockposition = this.blockPosition();
@@ -3460,7 +3447,7 @@ public abstract class LivingEntity extends Entity implements Attackable, net.caf
             Iterator iterator = map.entrySet().iterator();
 
             while (iterator.hasNext()) {
-                Entry<EquipmentSlot, ItemStack> entry = (Entry) iterator.next();
+                Map.Entry<EquipmentSlot, ItemStack> entry = (Map.Entry) iterator.next();
                 EquipmentSlot enumitemslot1 = (EquipmentSlot) entry.getKey();
                 ItemStack itemstack3 = (ItemStack) entry.getValue();
 
@@ -3648,7 +3635,7 @@ public abstract class LivingEntity extends Entity implements Attackable, net.caf
         Vec3 vec3d1 = new Vec3((double) this.xxa, (double) this.yya, (double) this.zza);
 
         if (this.hasEffect(MobEffects.SLOW_FALLING) || this.hasEffect(MobEffects.LEVITATION)) {
-            this.resetFallDistance();
+            //this.resetFallDistance(); // Cartridge
         }
 
         label104:
@@ -3916,7 +3903,7 @@ public abstract class LivingEntity extends Entity implements Attackable, net.caf
         super.rideTick();
         this.oRun = this.run;
         this.run = 0.0F;
-        this.resetFallDistance();
+        //this.resetFallDistance(); // Cartridge
     }
 
     @Override
diff --git a/src/main/java/net/minecraft/world/entity/projectile/ThrownEnderpearl.java b/src/main/java/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
index 334fa781e94d195c13dd78cdbe8f43b4b126ebd0..5b1d9900842ddadd42b83ed1e36bae0a3a60bd28 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
@@ -68,6 +68,7 @@ public class ThrownEnderpearl extends ThrowableItemProjectile {
 
                     if (entity instanceof ServerPlayer) {
                         ServerPlayer entityplayer = (ServerPlayer) entity;
+                        double originalY = entity.getY();
 
                         if (entityplayer.connection.isAcceptingMessages()) {
                             // CraftBukkit start
@@ -87,15 +88,18 @@ public class ThrownEnderpearl extends ThrowableItemProjectile {
                                 }
                             }
 
-                            // entity.changeDimension(new DimensionTransition(worldserver, this.position(), entity.getDeltaMovement(), entity.getYRot(), entity.getXRot(), DimensionTransition.DO_NOTHING)); // CraftBukkit - moved up
-                            entity.resetFallDistance();
                             entityplayer.resetCurrentImpulseContext();
-                            entity.hurt(this.damageSources().fall().customEventDamager(this), this.level().purpurConfig.enderPearlDamage); // CraftBukkit // Paper - fix DamageSource API // Purpur
+                            // Cartridge start - Make Ender Pearls increase fall-damage
+                            // entity.changeDimension(new DimensionTransition(worldserver, this.position(), entity.getDeltaMovement(), entity.getYRot(), entity.getXRot(), DimensionTransition.DO_NOTHING)); // CraftBukkit - moved up
+                            //entity.resetFallDistance();
+                            entityplayer.fallDistance += Math.max((float) (originalY - this.getY()), 0f);
+                            entityplayer.causeFallDamage(entityplayer.fallDistance, 1.0f, this.damageSources().fall().customEventDamager(this));
+                            // Cartridge end
                             this.playSound(worldserver, this.position());
                         }
                     } else {
                         entity.changeDimension(new DimensionTransition(worldserver, this.position(), entity.getDeltaMovement(), entity.getYRot(), entity.getXRot(), DimensionTransition.DO_NOTHING));
-                        entity.resetFallDistance();
+                        //entity.resetFallDistance(); // Cartridge
                         this.playSound(worldserver, this.position());
                     }
 
diff --git a/src/main/java/net/minecraft/world/entity/vehicle/Boat.java b/src/main/java/net/minecraft/world/entity/vehicle/Boat.java
index 5f3d58099d83fef4a9230e680b5e199d00eb7c7b..6c0f97e077f45fd2910613bbee6d898a2d649990 100644
--- a/src/main/java/net/minecraft/world/entity/vehicle/Boat.java
+++ b/src/main/java/net/minecraft/world/entity/vehicle/Boat.java
@@ -914,7 +914,7 @@ public class Boat extends VehicleEntity implements Leashable, VariantHolder<Boat
             if (onGround) {
                 if (this.fallDistance > 3.0F) {
                     if (this.status != Boat.Status.ON_LAND) {
-                        this.resetFallDistance();
+                        //this.resetFallDistance(); // Cartridge
                         return;
                     }
 
diff --git a/src/main/java/net/minecraft/world/item/ChorusFruitItem.java b/src/main/java/net/minecraft/world/item/ChorusFruitItem.java
index 9f59f63203807abc3802d3e2f060d90d0d754860..f0d1b05b176b4b799455ab5edfb4964fb39712b8 100644
--- a/src/main/java/net/minecraft/world/item/ChorusFruitItem.java
+++ b/src/main/java/net/minecraft/world/item/ChorusFruitItem.java
@@ -59,7 +59,8 @@ public class ChorusFruitItem extends Item {
                     }
 
                     world.playSound((Player) null, user.getX(), user.getY(), user.getZ(), soundeffect, soundcategory);
-                    user.resetFallDistance();
+                    //user.resetFallDistance(); // Cartridge
+                    user.fallDistance += (float) (vec3d.y - user.getY()); // Cartridge
                     break;
                 }
             }
diff --git a/src/main/java/net/minecraft/world/item/MaceItem.java b/src/main/java/net/minecraft/world/item/MaceItem.java
index e07d6f5589f5eaa95d9009b1b0af71c879b2dddb..8db60a614035ea4512b7d7537662de8e0b663f9a 100644
--- a/src/main/java/net/minecraft/world/item/MaceItem.java
+++ b/src/main/java/net/minecraft/world/item/MaceItem.java
@@ -75,9 +75,11 @@ public class MaceItem extends Item {
                 serverPlayer.currentImpulseImpactPos = serverPlayer.position();
             }
 
-            serverPlayer.setIgnoreFallDamageFromCurrentImpulse(true);
-            serverPlayer.setDeltaMovement(serverPlayer.getDeltaMovement().with(Direction.Axis.Y, 0.01F));
-            serverPlayer.connection.send(new ClientboundSetEntityMotionPacket(serverPlayer));
+            // Cartridge start - prevent mace from resetting fall-distance
+            serverPlayer.setIgnoreFallDamageFromCurrentImpulse(false);
+            //serverPlayer.setDeltaMovement(serverPlayer.getDeltaMovement().with(Direction.Axis.Y, 0.01F));
+            //serverPlayer.connection.send(new ClientboundSetEntityMotionPacket(serverPlayer));
+            // Cartridge end - prevent mace from resetting fall-distance
             if (target.onGround()) {
                 serverPlayer.setSpawnExtraParticlesOnFall(true);
                 SoundEvent soundEvent = serverPlayer.fallDistance > 5.0F ? SoundEvents.MACE_SMASH_GROUND_HEAVY : SoundEvents.MACE_SMASH_GROUND;
@@ -100,7 +102,8 @@ public class MaceItem extends Item {
     public void postHurtEnemy(ItemStack stack, LivingEntity target, LivingEntity attacker) {
         stack.hurtAndBreak(1, attacker, EquipmentSlot.MAINHAND);
         if (canSmashAttack(attacker)) {
-            attacker.resetFallDistance();
+            canSmashAttack(attacker);
+            //attacker.resetFallDistance(); // Cartridge
         }
     }
 
diff --git a/src/main/java/net/minecraft/world/level/ClipContext.java b/src/main/java/net/minecraft/world/level/ClipContext.java
index 28bf0518f7d17099d7e4990defbeda6757b4477c..5f0a99a9a500610132a30a628c768f2d1a3588b4 100644
--- a/src/main/java/net/minecraft/world/level/ClipContext.java
+++ b/src/main/java/net/minecraft/world/level/ClipContext.java
@@ -52,7 +52,7 @@ public class ClipContext {
     public static enum Block implements ClipContext.ShapeGetter {
 
         COLLIDER(BlockBehaviour.BlockStateBase::getCollisionShape), OUTLINE(BlockBehaviour.BlockStateBase::getShape), VISUAL(BlockBehaviour.BlockStateBase::getVisualShape), FALLDAMAGE_RESETTING((iblockdata, iblockaccess, blockposition, voxelshapecollision) -> {
-            return iblockdata.is(BlockTags.FALL_DAMAGE_RESETTING) ? Shapes.block() : Shapes.empty();
+            return Shapes.empty();
         });
 
         private final ClipContext.ShapeGetter shapeGetter;
diff --git a/src/main/java/net/minecraft/world/level/block/BedBlock.java b/src/main/java/net/minecraft/world/level/block/BedBlock.java
index 17c994a39a1b99cc7727e328ce7493d534247a21..b60a63041e68ed156ea8e36ab53c5c4d8f4b2997 100644
--- a/src/main/java/net/minecraft/world/level/block/BedBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/BedBlock.java
@@ -165,7 +165,7 @@ public class BedBlock extends HorizontalDirectionalBlock implements EntityBlock
     // CraftBukkit end
 
     public static boolean canSetSpawn(Level world) {
-        return world.dimensionType().bedWorks(); // Paper - actually check if the bed works
+        return false; // Cartridge
     }
 
     private boolean kickVillagerOutOfBed(Level world, BlockPos pos) {
@@ -181,7 +181,7 @@ public class BedBlock extends HorizontalDirectionalBlock implements EntityBlock
 
     @Override
     public void fallOn(Level world, BlockState state, BlockPos pos, Entity entity, float fallDistance) {
-        super.fallOn(world, state, pos, entity, fallDistance); // Purpur
+        super.fallOn(world, state, pos, entity, fallDistance); // Cartridge
     }
 
     @Override
diff --git a/src/main/java/net/minecraft/world/level/block/CoralBlock.java b/src/main/java/net/minecraft/world/level/block/CoralBlock.java
index bff97b7d3909f2ec9e58a341b901b3741927543f..aabfd7879b5f2c2acd7d61a336c1178e2956994a 100644
--- a/src/main/java/net/minecraft/world/level/block/CoralBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/CoralBlock.java
@@ -36,7 +36,8 @@ public class CoralBlock extends Block {
         return CoralBlock.CODEC;
     }
 
-    @Override
+    // Cartridge - Disable CoralBlocks from fading outside of water
+    /*@Override
     protected void tick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
         if (!this.scanForWater(world, pos)) {
             // CraftBukkit start
@@ -47,7 +48,7 @@ public class CoralBlock extends Block {
             world.setBlock(pos, this.deadBlock.defaultBlockState(), 2);
         }
 
-    }
+    }*/
 
     @Override
     protected BlockState updateShape(BlockState state, Direction direction, BlockState neighborState, LevelAccessor world, BlockPos pos, BlockPos neighborPos) {
diff --git a/src/main/java/net/minecraft/world/level/block/CoralFanBlock.java b/src/main/java/net/minecraft/world/level/block/CoralFanBlock.java
index 7701683000132479d8c8eaa77ac3f93e41a960be..843191d784a953873e8cf5aec5cc25d8bc851502 100644
--- a/src/main/java/net/minecraft/world/level/block/CoralFanBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/CoralFanBlock.java
@@ -37,7 +37,8 @@ public class CoralFanBlock extends BaseCoralFanBlock {
         this.tryScheduleDieTick(state, world, pos);
     }
 
-    @Override
+    // Cartridge - Disable CoralFanBlocks from fading outside of water
+    /*@Override
     protected void tick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
         if (!scanForWater(state, world, pos)) {
             // CraftBukkit start
@@ -48,7 +49,7 @@ public class CoralFanBlock extends BaseCoralFanBlock {
             world.setBlock(pos, (BlockState) this.deadBlock.defaultBlockState().setValue(CoralFanBlock.WATERLOGGED, false), 2);
         }
 
-    }
+    }*/
 
     @Override
     protected BlockState updateShape(BlockState state, Direction direction, BlockState neighborState, LevelAccessor world, BlockPos pos, BlockPos neighborPos) {
diff --git a/src/main/java/net/minecraft/world/level/block/CoralPlantBlock.java b/src/main/java/net/minecraft/world/level/block/CoralPlantBlock.java
index 4c4aa2a63be84ff614a3dc0db2864266755545a2..28462b2ab64941ddbbac856251ca2bab0f82ae3d 100644
--- a/src/main/java/net/minecraft/world/level/block/CoralPlantBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/CoralPlantBlock.java
@@ -42,7 +42,8 @@ public class CoralPlantBlock extends BaseCoralPlantTypeBlock {
         this.tryScheduleDieTick(state, world, pos);
     }
 
-    @Override
+    // Cartridge - Disable CoralPlantBlocks from fading outside of water
+    /*@Override
     protected void tick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
         if (!scanForWater(state, world, pos)) {
             // CraftBukkit start
@@ -53,7 +54,7 @@ public class CoralPlantBlock extends BaseCoralPlantTypeBlock {
             world.setBlock(pos, (BlockState) this.deadBlock.defaultBlockState().setValue(CoralPlantBlock.WATERLOGGED, false), 2);
         }
 
-    }
+    }*/
 
     @Override
     protected BlockState updateShape(BlockState state, Direction direction, BlockState neighborState, LevelAccessor world, BlockPos pos, BlockPos neighborPos) {
diff --git a/src/main/java/net/minecraft/world/level/block/CoralWallFanBlock.java b/src/main/java/net/minecraft/world/level/block/CoralWallFanBlock.java
index 924766117fcdf7285327e9aa1f89a630987855a8..e8376cb0ec2b79e7c94a11089ef1739def53165b 100644
--- a/src/main/java/net/minecraft/world/level/block/CoralWallFanBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/CoralWallFanBlock.java
@@ -37,18 +37,19 @@ public class CoralWallFanBlock extends BaseCoralWallFanBlock {
         this.tryScheduleDieTick(state, world, pos);
     }
 
-    @Override
-    protected void tick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
-        if (!scanForWater(state, world, pos)) {
-            // CraftBukkit start
-            if (org.bukkit.craftbukkit.event.CraftEventFactory.callBlockFadeEvent(world, pos, this.deadBlock.defaultBlockState().setValue(CoralWallFanBlock.WATERLOGGED, false).setValue(CoralWallFanBlock.FACING, state.getValue(CoralWallFanBlock.FACING))).isCancelled()) {
-                return;
-            }
-            // CraftBukkit end
-            world.setBlock(pos, (BlockState) ((BlockState) this.deadBlock.defaultBlockState().setValue(CoralWallFanBlock.WATERLOGGED, false)).setValue(CoralWallFanBlock.FACING, (Direction) state.getValue(CoralWallFanBlock.FACING)), 2);
-        }
-
-    }
+    // Cartridge - Disable CoralWallFanBlocks from fading outside of water
+//    @Override
+//    protected void tick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+//        if (!scanForWater(state, world, pos)) {
+//            // CraftBukkit start
+//            if (org.bukkit.craftbukkit.event.CraftEventFactory.callBlockFadeEvent(world, pos, this.deadBlock.defaultBlockState().setValue(CoralWallFanBlock.WATERLOGGED, false).setValue(CoralWallFanBlock.FACING, state.getValue(CoralWallFanBlock.FACING))).isCancelled()) {
+//                return;
+//            }
+//            // CraftBukkit end
+//            world.setBlock(pos, (BlockState) ((BlockState) this.deadBlock.defaultBlockState().setValue(CoralWallFanBlock.WATERLOGGED, false)).setValue(CoralWallFanBlock.FACING, (Direction) state.getValue(CoralWallFanBlock.FACING)), 2);
+//        }
+//
+//    }
 
     @Override
     protected BlockState updateShape(BlockState state, Direction direction, BlockState neighborState, LevelAccessor world, BlockPos pos, BlockPos neighborPos) {
diff --git a/src/main/java/net/minecraft/world/level/block/FarmBlock.java b/src/main/java/net/minecraft/world/level/block/FarmBlock.java
index d0ec0722496ed931b48c4e7076fddbb1ed36e111..12bb161c339302815ed613be8aa971ce4fe4e791 100644
--- a/src/main/java/net/minecraft/world/level/block/FarmBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/FarmBlock.java
@@ -108,7 +108,8 @@ public class FarmBlock extends Block {
 
     }
 
-    @Override
+    // Cartridge - Disable jumping on farmland converting to dirt
+    /*@Override
     public void fallOn(Level world, BlockState state, BlockPos pos, Entity entity, float fallDistance) {
         super.fallOn(world, state, pos, entity, fallDistance); // CraftBukkit - moved here as game rules / events shouldn't affect fall damage.
         if (!world.isClientSide && (world.purpurConfig.farmlandTrampleHeight >= 0D ? fallDistance >= world.purpurConfig.farmlandTrampleHeight : world.random.nextFloat() < fallDistance - 0.5F) && entity instanceof LivingEntity && (entity instanceof Player || world.purpurConfig.farmlandBypassMobGriefing || world.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) && entity.getBbWidth() * entity.getBbWidth() * entity.getBbHeight() > 0.512F) { // Purpur
@@ -149,7 +150,7 @@ public class FarmBlock extends Block {
         }
 
         // super.fallOn(world, iblockdata, blockposition, entity, f); // CraftBukkit - moved up
-    }
+    }*/
 
     public static void turnToDirt(@Nullable Entity entity, BlockState state, Level world, BlockPos pos) {
         // CraftBukkit start
diff --git a/src/main/java/net/minecraft/world/level/block/HayBlock.java b/src/main/java/net/minecraft/world/level/block/HayBlock.java
index 74971d90506aa253d5ee821b5390fb2551a3a393..0aaff453eb36630472e6aff932b1cdaa59f566c3 100644
--- a/src/main/java/net/minecraft/world/level/block/HayBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/HayBlock.java
@@ -23,6 +23,6 @@ public class HayBlock extends RotatedPillarBlock {
 
     @Override
     public void fallOn(Level world, BlockState state, BlockPos pos, Entity entity, float fallDistance) {
-        super.fallOn(world, state, pos, entity, fallDistance); // Purpur
+        entity.causeFallDamage(fallDistance, 1.0F, world.damageSources().fall()); // Cartridge
     }
 }
diff --git a/src/main/java/net/minecraft/world/level/block/HoneyBlock.java b/src/main/java/net/minecraft/world/level/block/HoneyBlock.java
index c6f7815b5fad3aad4635208aa2e5c6739e13cb45..403113eaa08963d764e71c08762b88c52c25de50 100644
--- a/src/main/java/net/minecraft/world/level/block/HoneyBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/HoneyBlock.java
@@ -53,7 +53,7 @@ public class HoneyBlock extends HalfTransparentBlock {
             world.broadcastEntityEvent(entity, (byte)54);
         }
 
-        if (entity.causeFallDamage(fallDistance, 0.2F, world.damageSources().fall())) {
+        if (entity.causeFallDamage(fallDistance, 1.0F, world.damageSources().fall())) { // Cartridge
             entity.playSound(this.soundType.getFallSound(), this.soundType.getVolume() * 0.5F, this.soundType.getPitch() * 0.75F);
         }
     }
@@ -100,7 +100,7 @@ public class HoneyBlock extends HalfTransparentBlock {
             entity.setDeltaMovement(new Vec3(vec3.x, -0.05, vec3.z));
         }
 
-        entity.resetFallDistance();
+        //entity.resetFallDistance(); // Cartridge
     }
 
     private void maybeDoSlideEffects(Level world, Entity entity) {
diff --git a/src/main/java/net/minecraft/world/level/block/IceBlock.java b/src/main/java/net/minecraft/world/level/block/IceBlock.java
index 87ce003dfeca975d8e6af26fd341f3abe66dc185..7d6d136c24c4e6ce999321add3dda18d978084e8 100644
--- a/src/main/java/net/minecraft/world/level/block/IceBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/IceBlock.java
@@ -55,13 +55,14 @@ public class IceBlock extends HalfTransparentBlock {
 
     }
 
-    @Override
-    protected void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
-        if (world.getBrightness(LightLayer.BLOCK, pos) > 11 - state.getLightBlock(world, pos)) {
-            this.melt(state, world, pos);
-        }
-
-    }
+    // Cartridge - prevent ice from melting
+//    @Override
+//    protected void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+//        if (world.getBrightness(LightLayer.BLOCK, pos) > 11 - state.getLightBlock(world, pos)) {
+//            this.melt(state, world, pos);
+//        }
+//
+//    }
 
     protected void melt(BlockState state, Level world, BlockPos pos) {
         // CraftBukkit start
diff --git a/src/main/java/net/minecraft/world/level/block/NoteBlock.java b/src/main/java/net/minecraft/world/level/block/NoteBlock.java
index 43dd972b374daa1072608f3a68e812e7fb733a2b..65cccce1b1548c134869e4b8035424ba1bd533a5 100644
--- a/src/main/java/net/minecraft/world/level/block/NoteBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/NoteBlock.java
@@ -66,32 +66,17 @@ public class NoteBlock extends Block {
 
     @Override
     public BlockState getStateForPlacement(BlockPlaceContext ctx) {
-        if (io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableNoteblockUpdates) return this.defaultBlockState(); // Paper - place without considering instrument
-        return this.setInstrument(ctx.getLevel(), ctx.getClickedPos(), this.defaultBlockState());
+        return this.defaultBlockState(); // Cartridge - place without considering instrument
     }
 
     @Override
     protected BlockState updateShape(BlockState state, Direction direction, BlockState neighborState, LevelAccessor world, BlockPos pos, BlockPos neighborPos) {
-        if (io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableNoteblockUpdates) return state; // Paper - prevent noteblock instrument from updating
-        boolean flag = direction.getAxis() == Direction.Axis.Y;
-
-        return flag ? this.setInstrument(world, pos, state) : super.updateShape(state, direction, neighborState, world, pos, neighborPos);
+        return state; // Cartridge
     }
 
     @Override
     protected void neighborChanged(BlockState state, Level world, BlockPos pos, Block sourceBlock, BlockPos sourcePos, boolean notify) {
-        if (io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableNoteblockUpdates) return; // Paper - prevent noteblock powered-state from updating
-        boolean flag1 = world.hasNeighborSignal(pos);
-
-        if (flag1 != (Boolean) state.getValue(NoteBlock.POWERED)) {
-            if (flag1) {
-                this.playNote((Entity) null, state, world, pos);
-                state = world.getBlockState(pos); // CraftBukkit - SPIGOT-5617: update in case changed in event
-            }
-
-            world.setBlock(pos, (BlockState) state.setValue(NoteBlock.POWERED, flag1), 3);
-        }
-
+        if (io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableNoteblockUpdates) return; // Cartridge - prevent noteblock powered-state from updating
     }
 
     private void playNote(@Nullable Entity entity, BlockState state, Level world, BlockPos pos) {
diff --git a/src/main/java/net/minecraft/world/level/block/PowderSnowBlock.java b/src/main/java/net/minecraft/world/level/block/PowderSnowBlock.java
index 5eac1a54398dfa5571b98fb6eefca9d2bf9b2793..aad457c0b89da4e251c452c3790ccebeea7ade31 100644
--- a/src/main/java/net/minecraft/world/level/block/PowderSnowBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/PowderSnowBlock.java
@@ -94,6 +94,7 @@ public class PowderSnowBlock extends Block implements BucketPickup {
 
     @Override
     public void fallOn(Level world, BlockState state, BlockPos pos, Entity entity, float fallDistance) {
+        entity.causeFallDamage(fallDistance, 1.0F, entity.damageSources().fall());
         if ((double) fallDistance >= 4.0D && entity instanceof LivingEntity entityliving) {
             LivingEntity.Fallsounds entityliving_a = entityliving.getFallSounds();
             SoundEvent soundeffect = (double) fallDistance < 7.0D ? entityliving_a.small() : entityliving_a.big();
diff --git a/src/main/java/net/minecraft/world/level/block/SlimeBlock.java b/src/main/java/net/minecraft/world/level/block/SlimeBlock.java
index fa9d3839b12e7f0485c91ae9a2f600314f41b893..a7033b3d563c42df339b78986c8d542b91c59fe4 100644
--- a/src/main/java/net/minecraft/world/level/block/SlimeBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/SlimeBlock.java
@@ -27,7 +27,7 @@ public class SlimeBlock extends HalfTransparentBlock {
         if (entity.isSuppressingBounce()) {
             super.fallOn(world, state, pos, entity, fallDistance);
         } else {
-            entity.causeFallDamage(fallDistance, 0.0F, world.damageSources().fall());
+            entity.causeFallDamage(fallDistance, 1.0F, world.damageSources().fall()); // Cartridge
         }
     }
 
@@ -57,5 +57,6 @@ public class SlimeBlock extends HalfTransparentBlock {
         }
 
         super.stepOn(world, pos, state, entity);
+        super.stepOn(world, pos, state, entity);
     }
 }
diff --git a/src/main/java/net/minecraft/world/level/block/SweetBerryBushBlock.java b/src/main/java/net/minecraft/world/level/block/SweetBerryBushBlock.java
index 6008c634c408c4eed563815da4d57b2eef69835c..51b2c3512c7619ca1fc91368e3b4d8a455cdbba7 100644
--- a/src/main/java/net/minecraft/world/level/block/SweetBerryBushBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/SweetBerryBushBlock.java
@@ -88,14 +88,15 @@ public class SweetBerryBushBlock extends BushBlock implements BonemealableBlock
         if (!new io.papermc.paper.event.entity.EntityInsideBlockEvent(entity.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(world, pos)).callEvent()) { return; } // Paper - Add EntityInsideBlockEvent
         if (entity instanceof LivingEntity && entity.getType() != EntityType.FOX && entity.getType() != EntityType.BEE) {
             entity.makeStuckInBlock(state, new Vec3(0.800000011920929D, 0.75D, 0.800000011920929D));
-            if (!world.isClientSide && (Integer) state.getValue(SweetBerryBushBlock.AGE) > 0 && (entity.xOld != entity.getX() || entity.zOld != entity.getZ())) {
-                double d0 = Math.abs(entity.getX() - entity.xOld);
-                double d1 = Math.abs(entity.getZ() - entity.zOld);
-
-                if (d0 >= 0.003000000026077032D || d1 >= 0.003000000026077032D) {
-                    entity.hurt(world.damageSources().sweetBerryBush().directBlock(world, pos), 1.0F); // CraftBukkit
-                }
-            }
+            // Cartridge - Disable BerryBush from damaging entities
+//            if (!world.isClientSide && (Integer) state.getValue(SweetBerryBushBlock.AGE) > 0 && (entity.xOld != entity.getX() || entity.zOld != entity.getZ())) {
+//                double d0 = Math.abs(entity.getX() - entity.xOld);
+//                double d1 = Math.abs(entity.getZ() - entity.zOld);
+//
+//                if (d0 >= 0.003000000026077032D || d1 >= 0.003000000026077032D) {
+//                    entity.hurt(world.damageSources().sweetBerryBush().directBlock(world, pos), 1.0F); // CraftBukkit
+//                }
+//            }
 
         }
     }
diff --git a/src/main/java/net/minecraft/world/level/block/TripWireBlock.java b/src/main/java/net/minecraft/world/level/block/TripWireBlock.java
index 6fe5be785423a35b6ff4e6206ca281b66845b979..8fd57e6228f876a0c81cd5f8e8ca89c83d55dfa4 100644
--- a/src/main/java/net/minecraft/world/level/block/TripWireBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/TripWireBlock.java
@@ -67,96 +67,61 @@ public class TripWireBlock extends Block {
 
     @Override
     public BlockState getStateForPlacement(BlockPlaceContext ctx) {
-        if (io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableTripwireUpdates) return this.defaultBlockState(); // Paper - place tripwire without updating
-        Level world = ctx.getLevel();
-        BlockPos blockposition = ctx.getClickedPos();
-
-        return (BlockState) ((BlockState) ((BlockState) ((BlockState) this.defaultBlockState().setValue(TripWireBlock.NORTH, this.shouldConnectTo(world.getBlockState(blockposition.north()), Direction.NORTH))).setValue(TripWireBlock.EAST, this.shouldConnectTo(world.getBlockState(blockposition.east()), Direction.EAST))).setValue(TripWireBlock.SOUTH, this.shouldConnectTo(world.getBlockState(blockposition.south()), Direction.SOUTH))).setValue(TripWireBlock.WEST, this.shouldConnectTo(world.getBlockState(blockposition.west()), Direction.WEST));
+        return this.defaultBlockState(); // Cartridge - place tripwire without updating
     }
 
     @Override
     protected BlockState updateShape(BlockState state, Direction direction, BlockState neighborState, LevelAccessor world, BlockPos pos, BlockPos neighborPos) {
-        if (io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableTripwireUpdates) return state; // Paper - prevent tripwire from updating
-        return direction.getAxis().isHorizontal() ? (BlockState) state.setValue((Property) TripWireBlock.PROPERTY_BY_DIRECTION.get(direction), this.shouldConnectTo(neighborState, direction)) : super.updateShape(state, direction, neighborState, world, pos, neighborPos);
+        return state; // Cartridge - prevent tripwire from updating
     }
 
+    // Cartridge
     @Override
     protected void onPlace(BlockState state, Level world, BlockPos pos, BlockState oldState, boolean notify) {
-        if (io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableTripwireUpdates) return; // Paper - prevent adjacent tripwires from updating
-        if (!oldState.is(state.getBlock())) {
-            this.updateSource(world, pos, state);
-        }
+//        if (io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableTripwireUpdates) return; // Paper - prevent adjacent tripwires from updating
+//        if (!oldState.is(state.getBlock())) {
+//            this.updateSource(world, pos, state);
+//        }
     }
 
+    // Cartridge
     @Override
     protected void onRemove(BlockState state, Level world, BlockPos pos, BlockState newState, boolean moved) {
-        if (io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableTripwireUpdates) return; // Paper - prevent adjacent tripwires from updating
-        if (!moved && !state.is(newState.getBlock())) {
-            this.updateSource(world, pos, (BlockState) state.setValue(TripWireBlock.POWERED, true));
-        }
+//        if (io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableTripwireUpdates) return; // Paper - prevent adjacent tripwires from updating
+//        if (!moved && !state.is(newState.getBlock())) {
+//            this.updateSource(world, pos, (BlockState) state.setValue(TripWireBlock.POWERED, true));
+//        }
     }
 
     @Override
     public BlockState playerWillDestroy(Level world, BlockPos pos, BlockState state, Player player) {
-        if (io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableTripwireUpdates) return state; // Paper - prevent disarming tripwires
-        if (!world.isClientSide && !player.getMainHandItem().isEmpty() && player.getMainHandItem().is(Items.SHEARS)) {
-            world.setBlock(pos, (BlockState) state.setValue(TripWireBlock.DISARMED, true), 4);
-            world.gameEvent((Entity) player, (Holder) GameEvent.SHEAR, pos);
-        }
-
-        return super.playerWillDestroy(world, pos, state, player);
+        return state; // Cartridge - prevent disarming tripwires
     }
 
     private void updateSource(Level world, BlockPos pos, BlockState state) {
-        if (io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableTripwireUpdates) return; // Paper - prevent adjacent tripwires from updating
-        Direction[] aenumdirection = new Direction[]{Direction.SOUTH, Direction.WEST};
-        int i = aenumdirection.length;
-        int j = 0;
-
-        while (j < i) {
-            Direction enumdirection = aenumdirection[j];
-            int k = 1;
-
-            while (true) {
-                if (k < 42) {
-                    BlockPos blockposition1 = pos.relative(enumdirection, k);
-                    BlockState iblockdata1 = world.getBlockState(blockposition1);
-
-                    if (iblockdata1.is(this.hook)) {
-                        if (iblockdata1.getValue(TripWireHookBlock.FACING) == enumdirection.getOpposite()) {
-                            TripWireHookBlock.calculateState(world, blockposition1, iblockdata1, false, true, k, state);
-                        }
-                    } else if (iblockdata1.is((Block) this)) {
-                        ++k;
-                        continue;
-                    }
-                }
-
-                ++j;
-                break;
-            }
-        }
-
-    }
-
-    @Override
-    protected void entityInside(BlockState state, Level world, BlockPos pos, Entity entity) {
-        if (io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableTripwireUpdates) return; // Paper - prevent tripwires from detecting collision
-        if (!new io.papermc.paper.event.entity.EntityInsideBlockEvent(entity.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(world, pos)).callEvent()) { return; } // Paper - Add EntityInsideBlockEvent
-        if (!world.isClientSide) {
-            if (!(Boolean) state.getValue(TripWireBlock.POWERED)) {
-                this.checkPressed(world, pos);
-            }
-        }
-    }
-
-    @Override
-    protected void tick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
-        if (io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableTripwireUpdates) return; // Paper - prevent tripwire pressed check
-        if ((Boolean) world.getBlockState(pos).getValue(TripWireBlock.POWERED)) {
-            this.checkPressed(world, pos);
-        }
-    }
+        if (io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableTripwireUpdates) return; // Cartridge - prevent adjacent tripwires from updating
+    }
+
+    // Cartridge - start
+//    @Override
+//    protected void entityInside(BlockState state, Level world, BlockPos pos, Entity entity) {
+//        if (io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableTripwireUpdates) return; // Paper - prevent tripwires from detecting collision
+//        if (!new io.papermc.paper.event.entity.EntityInsideBlockEvent(entity.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(world, pos)).callEvent()) { return; } // Paper - Add EntityInsideBlockEvent
+//        if (!world.isClientSide) {
+//            if (!(Boolean) state.getValue(TripWireBlock.POWERED)) {
+//                this.checkPressed(world, pos);
+//            }
+//        }
+//    }
+
+//    @Override
+//    protected void tick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+//        if (io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableTripwireUpdates) return; // Paper - prevent tripwire pressed check
+//        if ((Boolean) world.getBlockState(pos).getValue(TripWireBlock.POWERED)) {
+//            this.checkPressed(world, pos);
+//        }
+//    }
+    // Cartridge - end
 
     private void checkPressed(Level world, BlockPos pos) {
         BlockState iblockdata = world.getBlockState(pos);
diff --git a/src/main/java/net/minecraft/world/level/block/TurtleEggBlock.java b/src/main/java/net/minecraft/world/level/block/TurtleEggBlock.java
index 08ba90f760abb9fb62311dddd7b5bdbd0d9518d7..bdf4481e51c05be447e5e4ad4d659ea63e03e0a2 100644
--- a/src/main/java/net/minecraft/world/level/block/TurtleEggBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/TurtleEggBlock.java
@@ -66,14 +66,15 @@ public class TurtleEggBlock extends Block {
         super.stepOn(world, pos, state, entity);
     }
 
-    @Override
-    public void fallOn(Level world, BlockState state, BlockPos pos, Entity entity, float fallDistance) {
-        if (!(entity instanceof Zombie)) {
-            this.destroyEgg(world, state, pos, entity, 3);
-        }
-
-        super.fallOn(world, state, pos, entity, fallDistance);
-    }
+    // Cartridge - Disable eggs from breaking
+//    @Override
+//    public void fallOn(Level world, BlockState state, BlockPos pos, Entity entity, float fallDistance) {
+//        if (!(entity instanceof Zombie)) {
+//            this.destroyEgg(world, state, pos, entity, 3);
+//        }
+//
+//        super.fallOn(world, state, pos, entity, fallDistance);
+//    }
 
     private void destroyEgg(Level world, BlockState state, BlockPos pos, Entity entity, int inverseChance) {
         if (this.canDestroyEgg(world, entity)) {
@@ -82,6 +83,7 @@ public class TurtleEggBlock extends Block {
                 org.bukkit.event.Cancellable cancellable;
                 if (entity instanceof Player) {
                     cancellable = CraftEventFactory.callPlayerInteractEvent((Player) entity, org.bukkit.event.block.Action.PHYSICAL, pos, null, null, null);
+                    return;
                 } else {
                     cancellable = new EntityInteractEvent(entity.getBukkitEntity(), CraftBlock.at(world, pos));
                     world.getCraftServer().getPluginManager().callEvent((EntityInteractEvent) cancellable);
diff --git a/src/main/java/net/minecraft/world/level/material/FlowingFluid.java b/src/main/java/net/minecraft/world/level/material/FlowingFluid.java
index e3b23cb01aa25afcef32ce87bb4aafa6e418f190..0db2b8a81aa1b5333870e26c8a78056707e8fe52 100644
--- a/src/main/java/net/minecraft/world/level/material/FlowingFluid.java
+++ b/src/main/java/net/minecraft/world/level/material/FlowingFluid.java
@@ -534,38 +534,39 @@ public abstract class FlowingFluid extends Fluid {
         return this.getTickDelay(world);
     }
 
-    @Override
-    public void tick(Level world, BlockPos pos, FluidState state) {
-        if (!state.isSource()) {
-            FluidState fluid1 = this.getNewLiquid(world, pos, world.getBlockState(pos));
-            int i = this.getSpreadDelay(world, pos, state, fluid1);
-
-            if (fluid1.isEmpty()) {
-                state = fluid1;
-                // CraftBukkit start
-                FluidLevelChangeEvent event = CraftEventFactory.callFluidLevelChangeEvent(world, pos, Blocks.AIR.defaultBlockState());
-                if (event.isCancelled()) {
-                    return;
-                }
-                world.setBlock(pos, ((CraftBlockData) event.getNewData()).getState(), 3);
-                // CraftBukkit end
-            } else if (!fluid1.equals(state)) {
-                state = fluid1;
-                BlockState iblockdata = fluid1.createLegacyBlock();
-                // CraftBukkit start
-                FluidLevelChangeEvent event = CraftEventFactory.callFluidLevelChangeEvent(world, pos, iblockdata);
-                if (event.isCancelled()) {
-                    return;
-                }
-                world.setBlock(pos, ((CraftBlockData) event.getNewData()).getState(), 2);
-                // CraftBukkit end
-                world.scheduleTick(pos, fluid1.getType(), i);
-                world.updateNeighborsAt(pos, iblockdata.getBlock());
-            }
-        }
-
-        this.spread(world, pos, state);
-    }
+    // Cartridge - Disable liquid flow
+//    @Override
+//    public void tick(Level world, BlockPos pos, FluidState state) {
+//        if (!state.isSource()) {
+//            FluidState fluid1 = this.getNewLiquid(world, pos, world.getBlockState(pos));
+//            int i = this.getSpreadDelay(world, pos, state, fluid1);
+//
+//            if (fluid1.isEmpty()) {
+//                state = fluid1;
+//                // CraftBukkit start
+//                FluidLevelChangeEvent event = CraftEventFactory.callFluidLevelChangeEvent(world, pos, Blocks.AIR.defaultBlockState());
+//                if (event.isCancelled()) {
+//                    return;
+//                }
+//                world.setBlock(pos, ((CraftBlockData) event.getNewData()).getState(), 3);
+//                // CraftBukkit end
+//            } else if (!fluid1.equals(state)) {
+//                state = fluid1;
+//                BlockState iblockdata = fluid1.createLegacyBlock();
+//                // CraftBukkit start
+//                FluidLevelChangeEvent event = CraftEventFactory.callFluidLevelChangeEvent(world, pos, iblockdata);
+//                if (event.isCancelled()) {
+//                    return;
+//                }
+//                world.setBlock(pos, ((CraftBlockData) event.getNewData()).getState(), 2);
+//                // CraftBukkit end
+//                world.scheduleTick(pos, fluid1.getType(), i);
+//                world.updateNeighborsAt(pos, iblockdata.getBlock());
+//            }
+//        }
+//
+//        this.spread(world, pos, state);
+//    }
 
     protected static int getLegacyLevel(FluidState state) {
         return state.isSource() ? 0 : 8 - Math.min(state.getAmount(), 8) + ((Boolean) state.getValue(FlowingFluid.FALLING) ? 8 : 0);
diff --git a/src/main/resources/data/minecraft/tags/block/fall_damage_resetting.json b/src/main/resources/data/minecraft/tags/block/fall_damage_resetting.json
index a2882b9f161cf441c3ffc44bf89fa544725146b5..83c265b0fc31717d9e983b1ff29f09ff1902b0c3 100644
--- a/src/main/resources/data/minecraft/tags/block/fall_damage_resetting.json
+++ b/src/main/resources/data/minecraft/tags/block/fall_damage_resetting.json
@@ -1,7 +1,4 @@
 {
   "values": [
-    "#minecraft:climbable",
-    "minecraft:sweet_berry_bush",
-    "minecraft:cobweb"
   ]
 }
