From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Boy <sivertpaulsen2@gmail.com>
Date: Mon, 17 Jun 2024 12:54:41 +0200
Subject: [PATCH] Enchantment tweaks


diff --git a/src/main/java/net/minecraft/data/tags/VanillaEnchantmentTagsProvider.java b/src/main/java/net/minecraft/data/tags/VanillaEnchantmentTagsProvider.java
index bf112d49c13b126464401014ebbec0afd65c5034..6ce88860056d6abefc8f4e0b54d553577c618435 100644
--- a/src/main/java/net/minecraft/data/tags/VanillaEnchantmentTagsProvider.java
+++ b/src/main/java/net/minecraft/data/tags/VanillaEnchantmentTagsProvider.java
@@ -84,7 +84,7 @@ public class VanillaEnchantmentTagsProvider extends EnchantmentTagsProvider {
             .add(
                 Enchantments.PROTECTION,
                 Enchantments.FIRE_PROTECTION,
-                Enchantments.FEATHER_FALLING,
+                //Enchantments.FEATHER_FALLING, // Cartridge - Remove Feather Falling
                 Enchantments.BLAST_PROTECTION,
                 Enchantments.PROJECTILE_PROTECTION,
                 Enchantments.RESPIRATION,
diff --git a/src/main/java/net/minecraft/world/inventory/EnchantmentMenu.java b/src/main/java/net/minecraft/world/inventory/EnchantmentMenu.java
index e59a38d051179753fa9b29e8d746e25e162cac3e..748a34b0cbb2bb44d28c4aa79609f59db217f549 100644
--- a/src/main/java/net/minecraft/world/inventory/EnchantmentMenu.java
+++ b/src/main/java/net/minecraft/world/inventory/EnchantmentMenu.java
@@ -4,6 +4,8 @@ import com.mojang.datafixers.util.Pair;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Optional;
+
+import io.papermc.paper.registry.keys.tags.EnchantmentTagKeys;
 import net.minecraft.Util;
 import net.minecraft.advancements.CriteriaTriggers;
 import net.minecraft.core.BlockPos;
@@ -17,6 +19,7 @@ import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.sounds.SoundEvents;
 import net.minecraft.sounds.SoundSource;
 import net.minecraft.stats.Stats;
+import net.minecraft.tags.BlockTags;
 import net.minecraft.tags.EnchantmentTags;
 import net.minecraft.util.RandomSource;
 import net.minecraft.world.Container;
diff --git a/src/main/resources/data/minecraft/enchantment/feather_falling.json b/src/main/resources/data/minecraft/enchantment/feather_falling.json
index ff0e5cff695c48e529b74987ae909f89bb0f7829..3e4a1d50de6e8e5a5d71d602c9977a678400ae10 100644
--- a/src/main/resources/data/minecraft/enchantment/feather_falling.json
+++ b/src/main/resources/data/minecraft/enchantment/feather_falling.json
@@ -11,7 +11,7 @@
           "value": {
             "type": "minecraft:linear",
             "base": 3.0,
-            "per_level_above_first": 3.0
+            "per_level_above_first": 1.5
           }
         },
         "requirements": {
diff --git a/src/main/resources/data/minecraft/tags/enchantment/in_enchanting_table.json b/src/main/resources/data/minecraft/tags/enchantment/in_enchanting_table.json
index 2c8ec7f3a56038ee660c225f7b8876108f786ef6..cb79774ae3c28d90b24d7e664d62559db3fe842e 100644
--- a/src/main/resources/data/minecraft/tags/enchantment/in_enchanting_table.json
+++ b/src/main/resources/data/minecraft/tags/enchantment/in_enchanting_table.json
@@ -1,5 +1,6 @@
 {
   "values": [
-    "#minecraft:non_treasure"
+    "#minecraft:non_treasure",
+    "minecraft:mending"
   ]
 }
