From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Boy <sivertpaulsen2@gmail.com>
Date: Tue, 18 Jun 2024 19:57:06 +0200
Subject: [PATCH] Datacomponent fixes


diff --git a/src/main/java/io/papermc/paper/component/item/ComponentTypesBridgesImpl.java b/src/main/java/io/papermc/paper/component/item/ComponentTypesBridgesImpl.java
index 2e0996b652ecc01d4f53963e73340dc2828aecab..071df718283271df1a148e9ad4654da70e58d4c0 100644
--- a/src/main/java/io/papermc/paper/component/item/ComponentTypesBridgesImpl.java
+++ b/src/main/java/io/papermc/paper/component/item/ComponentTypesBridgesImpl.java
@@ -77,8 +77,8 @@ public final class ComponentTypesBridgesImpl implements ComponentTypesBridge {
     }
 
     @Override
-    public BundleContents.Builder mapItemColor() {
-        return new PaperBundleContents.BuilderImpl();
+    public MapItemColor.Builder mapItemColor() {
+        return new PaperMapItemColor.BuilderImpl();
     }
 
     @Override
diff --git a/src/main/java/io/papermc/paper/component/item/PaperJukeboxPlayable.java b/src/main/java/io/papermc/paper/component/item/PaperJukeboxPlayable.java
index dbf25090748b0c6b675a4681f0d0c1b826f6757b..95f80cc2aa94022ea3a18748f3a77d40d82ab05f 100644
--- a/src/main/java/io/papermc/paper/component/item/PaperJukeboxPlayable.java
+++ b/src/main/java/io/papermc/paper/component/item/PaperJukeboxPlayable.java
@@ -3,6 +3,11 @@ package io.papermc.paper.component.item;
 import com.google.common.base.Preconditions;
 import java.util.ArrayList;
 import java.util.List;
+
+import io.papermc.paper.registry.RegistryAccess;
+import io.papermc.paper.registry.RegistryKey;
+import io.papermc.paper.registry.TypedKey;
+import net.kyori.adventure.key.Key;
 import net.minecraft.world.item.EitherHolder;
 import org.bukkit.JukeboxSong;
 import org.bukkit.attribute.Attribute;
@@ -37,6 +42,11 @@ public record PaperJukeboxPlayable(
         return new PaperJukeboxPlayable(this.impl.withTooltip(showInTooltip));
     }
 
+    @Override
+    public JukeboxSong jukeboxSong() {
+        return CraftJukeboxSong.minecraftHolderToBukkit(this.impl.song().holder().get());
+    }
+
     static final class BuilderImpl implements Builder {
 
         private JukeboxSong song = null;
