From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Boy <sivertpaulsen2@gmail.com>
Date: Tue, 18 Jun 2024 19:57:06 +0200
Subject: [PATCH] Datacomponent fixes


diff --git a/src/main/java/io/papermc/paper/component/item/ComponentTypesBridgesImpl.java b/src/main/java/io/papermc/paper/component/item/ComponentTypesBridgesImpl.java
index 2e0996b652ecc01d4f53963e73340dc2828aecab..071df718283271df1a148e9ad4654da70e58d4c0 100644
--- a/src/main/java/io/papermc/paper/component/item/ComponentTypesBridgesImpl.java
+++ b/src/main/java/io/papermc/paper/component/item/ComponentTypesBridgesImpl.java
@@ -77,8 +77,8 @@ public final class ComponentTypesBridgesImpl implements ComponentTypesBridge {
     }
 
     @Override
-    public BundleContents.Builder mapItemColor() {
-        return new PaperBundleContents.BuilderImpl();
+    public MapItemColor.Builder mapItemColor() {
+        return new PaperMapItemColor.BuilderImpl();
     }
 
     @Override
